// @bridgedai/actions-core/tests/core.test.ts
import {
  createSignalEnvelope,
  validateSignalEnvelope,
  workflowSignalKey,
  artifactSignalKey,
  deploymentSignalKey,
  ociSignalKey,
  redactSecrets,
  redactUnknown,
  dedupeSignalsByIdempotencyKey,
  githubContextToSignalSubject,
  EVIDENCE_HINTS,
  getActionsCapabilities,
  isDeploymentRegistrationSupported,
  getCapabilities,
} from '../src/index';
import { parseEmitDemoSignals, signalMetadataFromEmitDemo } from '../src/signal-metadata';
import { ingestSignals } from '../src/signal-ingest';
import { BridgedHttpClient } from '../src/bridgedai-client';
import { readWorkflowInput, resolveWorkflowInput } from '../src/validation-url';

describe('signal envelope', () => {
  it('creates valid envelope', () => {
    const env = createSignalEnvelope({
      organizationId: 'org_550e8400-e29b-41d4-a716-446655440000',
      signalType: 'workflow_run.started',
      idempotencyKey: 'key-1',
      subject: { repository: 'o/r', commitSha: 'abc' },
    });
    expect(validateSignalEnvelope(env).signalType).toBe('workflow_run.started');
  });

  it('rejects missing organizationId', () => {
    expect(() =>
      createSignalEnvelope({
        organizationId: '',
        signalType: 'commit.discovered',
        idempotencyKey: 'k',
      }),
    ).toThrow(/organizationId/);
  });
});

describe('idempotency keys', () => {
  const ctx = { ownerRepo: 'o/r', workflowRunId: '99', runAttempt: '1', jobName: 'build' };

  it('generates workflow signal key', () => {
    expect(workflowSignalKey(ctx, 'runner.used')).toBe('github-actions:o/r:99:1:build:runner.used');
  });

  it('generates artifact signal key with digest', () => {
    expect(workflowSignalKey(ctx, 'artifact.pushed', 'sha256:abc')).toContain('abc');
  });

  it('generates oci and deployment keys', () => {
    expect(ociSignalKey('ghcr.io', 'app', 'sha256:dead', 'container_image.pushed')).toContain('oci:');
    expect(deploymentSignalKey('prod', 'api', 'sha256:dead', 'dep-1', 'deployment.completed')).toContain(
      'deployment:',
    );
  });
});

describe('redaction', () => {
  it('redacts api keys from messages', () => {
    const key = 'brdg_test_sk_secretvalue';
    expect(redactSecrets(`token=${key}`, key)).not.toContain('secretvalue');
  });

  it('redacts sensitive object keys', () => {
    const out = redactUnknown({ api_key: 'x', name: 'safe' }) as Record<string, unknown>;
    expect(out.api_key).toBe('[REDACTED_FIELD]');
    expect(out.name).toBe('safe');
  });

  it('redacts github tokens in strings', () => {
    const out = redactUnknown('ghs_abcdefghijklmnopqrstuvwxyz') as string;
    expect(out).toBe('[REDACTED_GITHUB_TOKEN]');
  });
});

describe('signal deduplication', () => {
  it('dedupes by idempotency key', () => {
    const a = createSignalEnvelope({
      organizationId: 'org',
      signalType: 'commit.discovered',
      idempotencyKey: 'same-key',
      rawPayload: { v: 1 },
    });
    const b = createSignalEnvelope({
      organizationId: 'org',
      signalType: 'commit.discovered',
      idempotencyKey: 'same-key',
      rawPayload: { v: 2 },
    });
    const deduped = dedupeSignalsByIdempotencyKey([a, b]);
    expect(deduped).toHaveLength(1);
    expect(deduped[0]!.rawPayload.v).toBe(2);
  });
});

describe('github context subject', () => {
  it('maps workflow context to signal subject', () => {
    const subject = githubContextToSignalSubject({
      repository: 'o/r',
      commitSha: 'abc123',
      workflowRunId: '42',
      runAttempt: '1',
      jobName: 'test',
    });
    expect(subject.repository).toBe('o/r');
    expect(subject.workflowRunId).toBe('42');
  });
});

describe('capabilities', () => {
  it('exports static contract', () => {
    const cap = getActionsCapabilities();
    expect(cap.releaseGate.supported).toBe(true);
    expect(cap.deploymentRegistration.supported).toBe(false);
    expect(isDeploymentRegistrationSupported()).toBe(false);
  });

  it('getCapabilities returns static contract', async () => {
    const cap = await getCapabilities({} as never);
    expect((cap as { ingestion?: { sbom?: boolean } }).ingestion?.sbom).toBe(true);
  });
});

describe('evidence hints', () => {
  it('exports canonical hint ids', () => {
    expect(EVIDENCE_HINTS.sbom).toBe('sbom');
    expect(EVIDENCE_HINTS.buildMetadata).toBe('build_metadata');
  });
});

describe('ingestSignals', () => {
  const signal = createSignalEnvelope({
    organizationId: 'org_550e8400-e29b-41d4-a716-446655440000',
    signalType: 'workflow_run.completed',
    idempotencyKey: 'k1',
    subject: { repository: 'o/r', artifactDigest: 'sha256:' + 'b'.repeat(64) },
  });

  it('uses native /v1/signals/ingest when available', async () => {
    const client = {
      requestJson: jest.fn(async () => ({ status: 'accepted', failed: 0, evidenceCreated: 1 })),
    } as unknown as BridgedHttpClient;
    const res = await ingestSignals(client, { orgId: 'org_550e8400-e29b-41d4-a716-446655440000', signals: [signal] });
    expect(res.ingestPath).toBe('native');
    expect(client.requestJson).toHaveBeenCalledWith(
      expect.objectContaining({ path: '/v1/signals/ingest', method: 'POST' }),
    );
  });

  it('falls back to artifact_metadata on 404', async () => {
    const client = Object.assign(new BridgedHttpClient({ baseUrl: 'https://api.test', apiKey: 'k', orgId: 'org_550e8400-e29b-41d4-a716-446655440000' }), {
      requestJson: jest.fn(async (opts: { path: string }) => {
        if (opts.path === '/v1/signals/ingest') {
          throw Object.assign(new Error('not found'), { status: 404 });
        }
        return { ok: true, parsed: true };
      }),
    });
    const res = await ingestSignals(client, { orgId: 'org_550e8400-e29b-41d4-a716-446655440000', signals: [signal] });
    expect(res.ingestPath).toBe('artifact_metadata');
    expect(res.uploaded).toBe(1);
  });
});

describe('signal metadata', () => {
  it('tags demo vs production from emit-demo-signals', () => {
    expect(signalMetadataFromEmitDemo('true')).toEqual({ dataSource: 'demo' });
    expect(signalMetadataFromEmitDemo('false')).toEqual({ dataSource: 'production', emitDemo: false });
    expect(parseEmitDemoSignals(undefined)).toBe(false);
  });
});

describe('readWorkflowInput', () => {
  const prev: Record<string, string | undefined> = {};

  beforeEach(() => {
    for (const k of ['INPUT_API-KEY', 'INPUT_API_KEY']) {
      prev[k] = process.env[k];
      delete process.env[k];
    }
  });

  afterEach(() => {
    for (const [k, v] of Object.entries(prev)) {
      if (v === undefined) delete process.env[k];
      else process.env[k] = v;
    }
  });

  it('reads GitHub runner env names with hyphens', () => {
    process.env['INPUT_API-KEY'] = 'from-github-runner';
    expect(readWorkflowInput('api-key')).toBe('from-github-runner');
  });

  it('falls back to legacy underscore env names', () => {
    process.env.INPUT_API_KEY = 'from-legacy-tests';
    expect(readWorkflowInput('api-key')).toBe('from-legacy-tests');
  });

  it('resolveWorkflowInput falls back to process.env', () => {
    process.env.BRIDGED_API_KEY = 'from-env-fallback';
    expect(resolveWorkflowInput('api-key', 'BRIDGED_API_KEY')).toBe('from-env-fallback');
    delete process.env.BRIDGED_API_KEY;
  });
});

describe('BridgedHttpClient bearer auth', () => {
  it('accepts accessToken without apiKey', () => {
    const client = new BridgedHttpClient({
      baseUrl: 'https://api.test',
      orgId: 'org_550e8400-e29b-41d4-a716-446655440000',
      accessToken: 'jwt-token',
    });
    expect(client.baseUrl).toBe('https://api.test');
  });

  it('requires apiKey or accessToken', () => {
    expect(() => new BridgedHttpClient({ baseUrl: 'https://api.test', orgId: 'org_550e8400-e29b-41d4-a716-446655440000' })).toThrow(
      /credentials/i,
    );
  });
});
