// @bridgedai/actions-core/src/auth-credentials.ts
/**
 * Resolve BridgedAI credentials for HTTP clients.
 * Org API keys (brdg_*) must use x-api-key, not Authorization: Bearer.
 */

/** Must match backend ApiKeyService.BRIDGED_API_KEY_PATTERN */
export const BRIDGED_API_KEY_PATTERN =
  /^brdg_(live|test)_sk_(ak_[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})_([A-Za-z0-9_-]{32,64})$/i;

export const MAX_BRIDGED_CREDENTIAL_LENGTH = 256;
export const ORG_ID_PATTERN = /^org_[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export type BridgedAuthCredential = {
  apiKey?: string;
  accessToken?: string;
};

export function isBridgedApiKey(value: string): boolean {
  const t = String(value ?? '').trim();
  return t.length > 0 && t.length <= MAX_BRIDGED_CREDENTIAL_LENGTH && BRIDGED_API_KEY_PATTERN.test(t);
}

export function isValidBridgedOrgId(orgId: string): boolean {
  return ORG_ID_PATTERN.test(String(orgId ?? '').trim());
}

/** Split a single secret into api-key vs JWT Bearer. */
export function splitBridgedCredential(token: string): BridgedAuthCredential {
  const t = String(token ?? '').trim();
  if (!t) return {};
  if (t.length > MAX_BRIDGED_CREDENTIAL_LENGTH) {
    throw new Error('BridgedAI credential exceeds maximum allowed length');
  }
  if (isBridgedApiKey(t)) return { apiKey: t };
  return { accessToken: t };
}

export type ResolveBridgedAuthInput = {
  apiKey?: string;
  accessToken?: string;
  bridgedaiToken?: string;
};

/**
 * Prefer explicit api-key; if access-token is a brdg_* key, route to x-api-key even when both are set.
 */
export function resolveBridgedAuth(input: ResolveBridgedAuthInput): BridgedAuthCredential {
  const apiKeyRaw = String(input.apiKey ?? '').trim();
  const tokenRaw = String(input.accessToken ?? input.bridgedaiToken ?? '').trim();

  if (apiKeyRaw.length > MAX_BRIDGED_CREDENTIAL_LENGTH || tokenRaw.length > MAX_BRIDGED_CREDENTIAL_LENGTH) {
    throw new Error('BridgedAI credential exceeds maximum allowed length');
  }

  if (apiKeyRaw) {
    if (isBridgedApiKey(apiKeyRaw)) return { apiKey: apiKeyRaw };
    return { accessToken: apiKeyRaw };
  }

  if (tokenRaw) return splitBridgedCredential(tokenRaw);
  return {};
}

export function requireBridgedAuth(input: ResolveBridgedAuthInput, message?: string): BridgedAuthCredential {
  const cred = resolveBridgedAuth(input);
  if (!cred.apiKey && !cred.accessToken) {
    throw new Error(message ?? 'Missing BridgedAI credentials (api-key or access-token required)');
  }
  return cred;
}
