// @bridgedai/actions-core/src/bridgedai-client.ts
import { createHash } from 'crypto';
import * as fs from 'fs';
import * as path from 'path';
import { isBridgedApiKey, isValidBridgedOrgId, MAX_BRIDGED_CREDENTIAL_LENGTH } from './auth-credentials';
import { normalizeBaseUrl } from './validation-url';
import { redactSecrets } from './redaction';
import { withRetry } from './retry';

export {
  normalizeBaseUrl,
  readWorkflowInput,
  resolveWorkflowInput,
  requireWorkflowInput,
} from './validation-url';

export interface BridgedHttpClientOptions {
  baseUrl: string;
  orgId: string;
  apiKey?: string;
  /** OIDC-exchanged BridgedAI JWT (Bearer auth). Mutually exclusive with apiKey for auth header selection. */
  accessToken?: string;
  timeoutMs?: number;
  preferAuthorizationApiKey?: boolean;
}

export interface JsonRequestOptions {
  method: 'GET' | 'POST';
  path: string;
  body?: unknown;
  allowRetry?: boolean;
}

export class BridgedApiError extends Error {
  readonly status: number;
  readonly code?: string;
  readonly detail?: string;
  readonly bodySnippet: string;

  constructor(opts: { status: number; message: string; code?: string; detail?: string; bodySnippet?: string }) {
    super(opts.message);
    this.name = 'BridgedApiError';
    this.status = opts.status;
    this.code = opts.code;
    this.detail = opts.detail;
    this.bodySnippet = opts.bodySnippet ?? '';
  }
}

function problemMessage(status: number, body: unknown, rawText: string): { msg: string; code?: string; detail?: string } {
  if (body && typeof body === 'object') {
    const o = body as Record<string, unknown>;
    const detail = typeof o.detail === 'string' ? o.detail : typeof o.message === 'string' ? o.message : undefined;
    const title = typeof o.title === 'string' ? o.title : undefined;
    const code = typeof o.code === 'string' ? o.code : typeof o.error === 'string' ? o.error : undefined;
    const msg = [title, detail].filter(Boolean).join(': ') || `HTTP ${status}`;
    return { msg, code, detail };
  }
  const t = rawText.length > 2000 ? `${rawText.slice(0, 2000)}…` : rawText;
  return { msg: t || `HTTP ${status}` };
}

export class BridgedHttpClient {
  readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly accessToken: string;
  private readonly orgId: string;
  private readonly timeoutMs: number;
  private readonly preferAuthorizationApiKey: boolean;

  constructor(opts: BridgedHttpClientOptions) {
    this.baseUrl = normalizeBaseUrl(opts.baseUrl);
    let apiKey = String(opts.apiKey ?? '').trim();
    let accessToken = String(opts.accessToken ?? '').trim();
    if (apiKey.length > MAX_BRIDGED_CREDENTIAL_LENGTH || accessToken.length > MAX_BRIDGED_CREDENTIAL_LENGTH) {
      throw new Error('BridgedAI credential exceeds maximum allowed length');
    }
    if (accessToken && isBridgedApiKey(accessToken)) {
      apiKey = accessToken;
      accessToken = '';
    }
    this.apiKey = apiKey;
    this.accessToken = accessToken;
    this.orgId = String(opts.orgId ?? '').trim();
    this.timeoutMs = opts.timeoutMs ?? 30_000;
    this.preferAuthorizationApiKey = opts.preferAuthorizationApiKey ?? false;
    if (!this.apiKey && !this.accessToken) {
      throw new Error('Missing BridgedAI credentials (apiKey or accessToken required)');
    }
    if (!this.orgId) throw new Error('Missing BridgedAI organization id');
    if (!isValidBridgedOrgId(this.orgId)) {
      throw new Error('Invalid BridgedAI organization id format (expected org_<uuid>)');
    }
  }

  /** Credential used for log redaction (never logged itself). */
  private redactionSecret(): string {
    return this.accessToken || this.apiKey;
  }

  private authHeaders(): Record<string, string> {
    if (this.accessToken) {
      return {
        Authorization: `Bearer ${this.accessToken}`,
        'x-org-id': this.orgId,
      };
    }
    if (this.preferAuthorizationApiKey) {
      return {
        Authorization: `ApiKey ${this.apiKey}`,
        'x-org-id': this.orgId,
      };
    }
    return {
      'x-api-key': this.apiKey,
      'x-org-id': this.orgId,
    };
  }

  async requestJson<T>(opts: JsonRequestOptions): Promise<T> {
    const url = `${this.baseUrl}${opts.path.startsWith('/') ? opts.path : `/${opts.path}`}`;
    const allowRetry = opts.allowRetry ?? opts.method === 'GET';

    return withRetry(
      async () => {
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), this.timeoutMs);
        try {
          const headers: Record<string, string> = {
            accept: 'application/json',
            ...this.authHeaders(),
          };
          const init: RequestInit = { method: opts.method, headers, signal: controller.signal };
          if (opts.body !== undefined && opts.method === 'POST') {
            headers['content-type'] = 'application/json';
            init.body = JSON.stringify(opts.body);
          }

          const res = await fetch(url, init);
          const text = await res.text();
          let parsed: unknown = undefined;
          if (text) {
            try {
              parsed = JSON.parse(text) as unknown;
            } catch {
              parsed = { raw: text };
            }
          }

          if (res.ok) return parsed as T;

          const { msg, code, detail } = problemMessage(res.status, parsed, text);
          const secret = this.redactionSecret();
          const snippet = redactSecrets(text.slice(0, 800), secret);
          let userMsg = redactSecrets(msg, secret);
          if (res.status === 422) {
            userMsg += ' (Unprocessable entity — verify JSON body matches the public CI API contract.)';
          }

          const retryable = res.status === 502 || res.status === 503 || res.status === 504;
          if (allowRetry && retryable) {
            throw Object.assign(new BridgedApiError({
              status: res.status,
              message: `BridgedAI API ${res.status}: ${userMsg}`,
              code,
              detail,
              bodySnippet: snippet,
            }), { retryable: true });
          }

          throw new BridgedApiError({
            status: res.status,
            message: `BridgedAI API ${res.status}: ${userMsg}`,
            code,
            detail,
            bodySnippet: snippet,
          });
        } catch (e) {
          if (e instanceof BridgedApiError) throw e;
          const aborted = e instanceof Error && e.name === 'AbortError';
          if (aborted) {
            throw new BridgedApiError({
              status: 0,
              message: `BridgedAI request timed out after ${this.timeoutMs}ms (${opts.method} ${opts.path})`,
            });
          }
          throw e;
        } finally {
          clearTimeout(timer);
        }
      },
      {
        maxAttempts: allowRetry ? 3 : 1,
        isRetryable: (e) => {
          if (e && typeof e === 'object' && 'retryable' in e) return !!(e as { retryable: boolean }).retryable;
          if (e instanceof BridgedApiError) return e.status === 502 || e.status === 503 || e.status === 504;
          return false;
        },
      },
    );
  }
}

export interface ReleaseGateEvaluateRequest {
  orgId: string;
  buildId?: string;
  repoFullName?: string;
  workflowRunId?: string;
  commitSha?: string;
  projectId?: string;
  artifactId?: string;
  artifactDigest?: string;
  branch?: string;
  environment?: string;
  policyId?: string;
}

export interface ReleaseGateEvaluateResponse {
  decision: 'allow' | 'warn' | 'block';
  decisionId: string;
  trustScore?: number;
  reasons: string[];
  findings: Array<{ code: string; message: string; severity?: string }>;
  evaluatedAt: string;
  policyId: string;
  buildId: string;
  artifactId?: string;
}

export async function evaluateReleaseGate(
  client: BridgedHttpClient,
  body: ReleaseGateEvaluateRequest,
): Promise<ReleaseGateEvaluateResponse> {
  return client.requestJson<ReleaseGateEvaluateResponse>({
    method: 'POST',
    path: '/v1/enforcement/release-gate/evaluate',
    body,
    allowRetry: false,
  });
}

export interface DeploymentGuardEvaluateResponse {
  decision: string;
  decisionId?: string;
  deploymentId?: string;
  artifactId?: string;
  policyStatus?: string;
  trustScore?: number;
  riskSeverity?: string;
  verificationStatus?: string;
  blockingReasons?: string[];
  warnings?: string[];
  findings?: Array<{ code?: string; severity?: string; message?: string; title?: string }>;
  dashboardUrl?: string;
  explanations?: string[];
  reasonCodes?: string[];
  enforcementMode?: string;
}

export async function evaluateDeploymentGuard(
  client: BridgedHttpClient,
  body: Record<string, unknown>,
): Promise<DeploymentGuardEvaluateResponse> {
  return client.requestJson<DeploymentGuardEvaluateResponse>({
    method: 'POST',
    path: '/v1/deployment-guard/evaluate',
    body,
    allowRetry: false,
  });
}

export interface RegisterDeploymentResponse {
  deployment?: { id?: string; artifact?: { id?: string } | null };
  idempotent?: boolean;
  reasonCodes?: string[];
}

export async function registerDeployment(
  client: BridgedHttpClient,
  body: Record<string, unknown>,
): Promise<RegisterDeploymentResponse> {
  return client.requestJson<RegisterDeploymentResponse>({
    method: 'POST',
    path: '/v1/deployments/register',
    body,
    allowRetry: false,
  });
}

export interface IngestSbomRequest {
  sbom: Record<string, unknown> | string;
  buildId?: string;
  source?: string;
  linkEvidence?: boolean;
  repoId?: string;
  repoFullName?: string;
  publishToGuac?: boolean;
}

export interface IngestResult {
  ok: boolean;
  parsed: boolean;
  guacPublished?: boolean;
  graphWritten?: boolean;
  evidenceEmitted?: number;
  evidenceId?: string;
  error?: string;
  status: string;
}

export async function ingestSbom(client: BridgedHttpClient, body: IngestSbomRequest): Promise<IngestResult> {
  return client.requestJson<IngestResult>({
    method: 'POST',
    path: '/v1/ingest/sbom',
    body,
    allowRetry: false,
  });
}

export interface IngestAttestationRequest {
  attestation: Record<string, unknown> | string;
  buildId?: string;
  projectId?: string;
  artifactId?: string;
  publishToGuac?: boolean;
}

export async function ingestAttestation(client: BridgedHttpClient, body: IngestAttestationRequest): Promise<IngestResult> {
  return client.requestJson<IngestResult>({
    method: 'POST',
    path: '/v1/ingest/attestation',
    body,
    allowRetry: false,
  });
}

export const uploadSbom = ingestSbom;
export const uploadAttestation = ingestAttestation;

export async function getCapabilities(_client: BridgedHttpClient): Promise<Record<string, unknown>> {
  void _client;
  const { getActionsCapabilities } = await import('./capabilities');
  return getActionsCapabilities() as unknown as Record<string, unknown>;
}

export interface BuildResolveRequest {
  orgId: string;
  buildId?: string;
  repoFullName?: string;
  workflowRunId?: string;
  commitSha?: string;
  branch?: string;
}

export interface BuildResolveResponse {
  buildId?: string;
  projectId?: string;
  repoFullName?: string;
  workflowRunId?: string;
  commitSha?: string;
  branch?: string;
  resolved?: boolean;
  [key: string]: unknown;
}

export async function resolveBuild(client: BridgedHttpClient, body: BuildResolveRequest): Promise<BuildResolveResponse> {
  return client.requestJson<BuildResolveResponse>({
    method: 'POST',
    path: '/v1/builds/resolve',
    body,
    allowRetry: false,
  });
}

/** CI resolve: always send workflowRunId and commitSha so the API can match legacy builds. */
export async function resolveBuildForCi(
  client: BridgedHttpClient,
  params: {
    orgId: string;
    repoFullName: string;
    workflowRunId?: string;
    commitSha?: string;
    branch?: string;
  },
): Promise<BuildResolveResponse> {
  return resolveBuild(client, {
    orgId: params.orgId,
    repoFullName: params.repoFullName,
    workflowRunId: params.workflowRunId,
    commitSha: params.commitSha,
    branch: params.branch,
  });
}

export interface BuildEnsureRequest {
  orgId: string;
  repoFullName: string;
  workflowRunId: string;
  commitSha: string;
  branch?: string;
  workflowName?: string;
  workflowPath?: string;
  status?: string;
}

export interface BuildEnsureResponse extends BuildResolveResponse {
  created?: boolean;
  ensured?: boolean;
}

/** Idempotent CI registration — creates build immediately without waiting for webhooks. */
export async function ensureBuild(client: BridgedHttpClient, body: BuildEnsureRequest): Promise<BuildEnsureResponse> {
  return client.requestJson<BuildEnsureResponse>({
    method: 'POST',
    path: '/v1/builds/ensure',
    body,
    allowRetry: false,
  });
}

export function extractBuildIdFromResolveResponse(res: BuildResolveResponse | Record<string, unknown>): string {
  if (!res || typeof res !== 'object') return '';
  const o = res as Record<string, unknown>;
  if (typeof o.buildId === 'string' && o.buildId) return o.buildId;
  if (typeof o.build_id === 'string' && o.build_id) return o.build_id;
  const b = o.build;
  if (b && typeof b === 'object') {
    const id = (b as Record<string, unknown>).id;
    if (typeof id === 'string' && id) return id;
  }
  return '';
}

export type EvidenceKind =
  | 'sbom'
  | 'intoto_statement'
  | 'slsa_provenance'
  | 'sigstore_bundle'
  | 'sigstore_signature'
  | 'rekor_entry'
  | 'artifact_metadata';

export async function ingestEvidence(client: BridgedHttpClient, body: Record<string, unknown>): Promise<IngestResult> {
  return client.requestJson<IngestResult>({
    method: 'POST',
    path: '/v1/ingest/evidence',
    body,
    allowRetry: false,
  });
}

export const uploadGenericEvidence = ingestEvidence;

export interface VerifySupplyChainBody {
  policyId?: string;
  offline?: boolean;
}

export type VerificationStatus = 'verified' | 'failed' | 'warning' | 'unsupported' | 'not_present';

export interface VerificationResult {
  artifactId?: string;
  buildId?: string;
  status: VerificationStatus;
  verified: boolean;
  findings: Array<{ code: string; title?: string; message: string; severity?: string; status?: string }>;
  verifiedAt?: string;
  verifier?: string;
  metadata?: Record<string, unknown>;
}

export async function verifyArtifactSupplyChain(
  client: BridgedHttpClient,
  artifactId: string,
  body: VerifySupplyChainBody,
): Promise<VerificationResult> {
  const urlPath = `/v1/artifacts/${encodeURIComponent(artifactId)}/verify-supply-chain`;
  return client.requestJson<VerificationResult>({
    method: 'POST',
    path: urlPath,
    body,
    allowRetry: false,
  });
}

export interface BuildSupplyChainVerificationResponse {
  buildId: string;
  evaluationId?: string;
  runs: unknown[];
  findings: unknown[];
  evaluatedAt?: string;
}

export async function getBuildSupplyChainVerification(
  client: BridgedHttpClient,
  buildId: string,
): Promise<BuildSupplyChainVerificationResponse> {
  const urlPath = `/v1/builds/${encodeURIComponent(buildId)}/supply-chain-verification`;
  return client.requestJson<BuildSupplyChainVerificationResponse>({
    method: 'GET',
    path: urlPath,
    allowRetry: true,
  });
}

export async function getCurrentUserMe(client: BridgedHttpClient): Promise<Record<string, unknown>> {
  return client.requestJson<Record<string, unknown>>({
    method: 'GET',
    path: '/v1/users/me',
    allowRetry: true,
  });
}

export function sha256HexOfBuffer(buf: Buffer): string {
  return createHash('sha256').update(buf).digest('hex');
}

export function sha256HexOfFile(filePath: string, fsMod: typeof fs = fs): string {
  return sha256HexOfBuffer(fsMod.readFileSync(filePath));
}

export function listFilesUnderDirectory(rootAbs: string, fsMod: typeof fs = fs): string[] {
  const root = path.resolve(rootAbs);
  const files: string[] = [];
  const walk = (dir: string) => {
    for (const ent of fsMod.readdirSync(dir, { withFileTypes: true })) {
      if (ent.name === '.git' && ent.isDirectory()) continue;
      const abs = path.join(dir, ent.name);
      if (ent.isDirectory()) walk(abs);
      else if (ent.isFile()) files.push(abs);
    }
  };
  walk(root);
  return files.sort((a, b) => a.localeCompare(b));
}

export function computeDirectoryManifestSha256(rootAbs: string, fsMod: typeof fs = fs): string {
  const root = path.resolve(rootAbs);
  const h = createHash('sha256');
  const files = listFilesUnderDirectory(root, fsMod);
  for (const abs of files) {
    const rel = path.relative(root, abs).split(path.sep).join('/');
    const fileHash = sha256HexOfFile(abs, fsMod);
    h.update(rel);
    h.update('\n');
    h.update(fileHash);
    h.update('\n');
  }
  return h.digest('hex');
}
