// @bridgedai/actions-core/src/capabilities.ts
/**
 * Static CI capabilities contract for BridgedAI GitHub Actions.
 * Previously served by GET /v1/actions/capabilities on the backend; now owned by the actions ecosystem.
 * @see private/actions-dev/docs/API_COMPATIBILITY.md
 */
export const BRIDGED_ACTIONS_CAPABILITIES = {
  schemaVersion: '2026-05-01',
  releaseGate: {
    supported: true,
    endpoint: '/v1/enforcement/release-gate/evaluate',
    buildResolution: ['buildId', 'repoFullName+workflowRunId', 'repoFullName+commitSha'],
    requiredPermission: 'release-gates:evaluate',
  },
  buildResolution: {
    supported: true,
    endpoint: '/v1/builds/resolve',
    requiredPermission: 'release-gates:evaluate',
  },
  ingestion: {
    sbom: true,
    attestation: true,
    genericEvidence: true,
    sbomLegacyBuildIdOptional: true,
    sbomEvidenceEmissionRequiresBuildId: true,
    sbomCiRequiresBuildIdWhen: ['source=github-actions', 'linkEvidence=true'],
    supportedKinds: [
      'sbom',
      'intoto_statement',
      'slsa_provenance',
      'sigstore_bundle',
      'sigstore_signature',
      'rekor_entry',
      'artifact_metadata',
    ],
    endpoints: {
      sbom: '/v1/ingest/sbom',
      attestation: '/v1/ingest/attestation',
      evidence: '/v1/ingest/evidence',
    },
    uploadPermissions: ['evidence:upload', 'projects:update'],
  },
  verification: {
    supplyChainStructural: true,
    sigstoreCrypto: false,
    rekorOnline: false,
    artifactVerifyEndpoint: '/v1/artifacts/:artifactId/verify-supply-chain',
    buildVerificationEndpoint: '/v1/builds/:buildId/supply-chain-verification',
  },
  deploymentRegistration: {
    supported: false,
    dryRunOnly: true,
  },
  signals: {
    /** Native POST /v1/signals/ingest when backend supports it; actions client falls back to evidence bridge. */
    nativeIngestEndpoint: true,
    nativeIngestPath: '/v1/signals/ingest',
    ingestViaEvidence: true,
    dedicatedIngestEndpoint: true,
    batchFormat: 'bridgedai-signal-batch-v1',
  },
} as const;

export type BridgedActionsCapabilities = typeof BRIDGED_ACTIONS_CAPABILITIES;

/** @deprecated Use getActionsCapabilities() — backend GET /v1/actions/capabilities removed. */
export function getActionsCapabilities(): Readonly<BridgedActionsCapabilities> {
  return BRIDGED_ACTIONS_CAPABILITIES;
}

export function isDeploymentRegistrationSupported(
  cap: Readonly<Record<string, unknown>> = BRIDGED_ACTIONS_CAPABILITIES as unknown as Record<string, unknown>,
): boolean {
  const dep = cap.deploymentRegistration as { supported?: boolean } | undefined;
  return dep?.supported === true;
}

export function isSigstoreCryptoSupported(
  cap: Readonly<Record<string, unknown>> = BRIDGED_ACTIONS_CAPABILITIES as unknown as Record<string, unknown>,
): boolean {
  const ver = cap.verification as { sigstoreCrypto?: boolean } | undefined;
  return ver?.sigstoreCrypto === true;
}
