// @bridgedai/actions-core/src/deployment-payload.ts
import type { GitHubWorkflowContext } from './github-context';
import { deploymentRegisteredIdempotencyKey } from './idempotency';
import { isSensitiveKey, redactUnknown } from './redaction';

export type DeploymentGuardMode = 'observe' | 'warn' | 'enforce';

export type DeploymentRegistrationInputs = {
  repository?: string;
  workflowRunId?: string;
  runAttempt?: string;
  commitSha?: string;
  branch?: string;
  ref?: string;
  actor?: string;
  environment: string;
  environmentType?: string;
  deploymentName?: string;
  serviceName?: string;
  deploymentUrl?: string;
  region?: string;
  cluster?: string;
  namespace?: string;
  artifactId?: string;
  artifactDigest?: string;
  imageName?: string;
  imageTag?: string;
  imageDigest?: string;
  packageName?: string;
  packageVersion?: string;
  buildId?: string;
  projectId?: string;
  repositoryId?: string;
  externalId?: string;
  releaseGateDecisionId?: string;
  releaseGateDecision?: string;
  deploymentGuardDecisionId?: string;
  correlationId?: string;
  jobId?: string;
  jobName?: string;
  runnerId?: string;
  status?: string;
  provider?: string;
  deploymentId?: string;
};

const SHA256 = /^sha256:([a-f0-9]{64})$/i;
const HEX64 = /^[a-f0-9]{64}$/i;

export function tryNormalizeDeploymentDigest(raw?: string): string | undefined {
  const s = String(raw ?? '').trim();
  if (!s) return undefined;
  const m = s.match(SHA256);
  if (m) return `sha256:${m[1]!.toLowerCase()}`;
  if (HEX64.test(s)) return `sha256:${s.toLowerCase()}`;
  return undefined;
}

/** Map workflow environment labels to backend deployment environment values. */
export function normalizeDeploymentEnvironment(raw: string): string {
  const v = String(raw ?? '').trim().toLowerCase();
  if (!v) return 'unknown';
  if (v === 'prod' || v === 'production') return 'production';
  if (v === 'stage' || v === 'staging') return 'staging';
  if (v === 'dev' || v === 'development') return 'development';
  if (v === 'preview' || v === 'preprod') return 'preview';
  return v;
}

export function inferEnvironmentType(environment: string, explicit?: string): string | undefined {
  const e = String(explicit ?? '').trim();
  if (e) return e;
  const norm = normalizeDeploymentEnvironment(environment);
  if (norm === 'production' || norm === 'staging' || norm === 'development' || norm === 'preview') {
    return norm;
  }
  return undefined;
}

/** Strip sensitive keys from deployment metadata before POST. */
export function sanitizeDeploymentMetadata(metadata: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(metadata)) {
    if (isSensitiveKey(k)) continue;
    out[k] = redactUnknown(v);
  }
  return out;
}

export function mapDeploymentStatus(raw?: string): string {
  const s = String(raw ?? 'succeeded').trim().toLowerCase();
  if (s === 'completed' || s === 'success' || s === 'succeeded') return 'succeeded';
  if (s === 'failed' || s === 'failure') return 'failed';
  if (s === 'rolled_back' || s === 'rollback') return 'rolled_back';
  if (s === 'pending' || s === 'in_progress' || s === 'cancelled') return s;
  return 'succeeded';
}

export type BuildDeploymentPayloadResult = {
  body: Record<string, unknown>;
  warnings: string[];
  digest?: string;
  mutableTagOnly: boolean;
  idempotencyKey?: string;
};

export function buildDeploymentRegistrationPayload(
  inputs: DeploymentRegistrationInputs,
  ghCtx?: GitHubWorkflowContext,
): BuildDeploymentPayloadResult {
  const warnings: string[] = [];
  const digest =
    tryNormalizeDeploymentDigest(inputs.artifactDigest) ?? tryNormalizeDeploymentDigest(inputs.imageDigest);
  const imageName = String(inputs.imageName ?? '').trim();
  const imageTag = String(inputs.imageTag ?? '').trim();
  const mutableTagOnly = !digest && !!(imageName && imageTag);
  if (mutableTagOnly) {
    warnings.push(
      'Deployment registered with mutable tag reference only; pass artifact-digest or image-digest for digest-first verification.',
    );
  }
  if (!digest && !mutableTagOnly && !inputs.artifactId) {
    warnings.push('No immutable digest supplied; artifact correlation may be incomplete.');
  }

  const repository = String(inputs.repository ?? ghCtx?.repository ?? '').trim();
  const workflowRunId = inputs.workflowRunId ?? ghCtx?.workflowRunId;
  const runAttempt = inputs.runAttempt ?? ghCtx?.runAttempt ?? '1';
  const environment = normalizeDeploymentEnvironment(inputs.environment);
  const deploymentName = String(inputs.deploymentName ?? inputs.serviceName ?? '').trim();
  const serviceName = String(inputs.serviceName ?? deploymentName ?? repository.split('/').pop() ?? '').trim();

  const idempotencyKey =
    repository && workflowRunId
      ? deploymentRegisteredIdempotencyKey({
          repository,
          workflowRunId,
          runAttempt,
          environment,
          artifactDigest: digest,
          imageDigest: digest,
          deploymentName: deploymentName || serviceName,
        })
      : undefined;

  const metadata = sanitizeDeploymentMetadata({
    registrationChannel: 'github-actions',
    correlationId: inputs.correlationId || undefined,
    idempotencyKey,
    deploymentName: deploymentName || undefined,
    deploymentUrl: inputs.deploymentUrl || undefined,
    environmentType: inferEnvironmentType(inputs.environment, inputs.environmentType),
    imageTag: imageTag || undefined,
    packageName: inputs.packageName || undefined,
    packageVersion: inputs.packageVersion || undefined,
    deploymentGuardDecisionId: inputs.deploymentGuardDecisionId || undefined,
    jobId: inputs.jobId || ghCtx?.workflowJobId || undefined,
    jobName: inputs.jobName || ghCtx?.jobName || undefined,
    runAttempt,
    projectId: inputs.projectId || undefined,
    repositoryId: inputs.repositoryId || ghCtx?.repositoryId || undefined,
    owner: ghCtx?.repositoryOwner,
    repo: repository.includes('/') ? repository.split('/', 2)[1] : undefined,
    workflow: ghCtx?.workflowName,
    workflowRef: ghCtx?.workflowRef,
    actor: inputs.actor || ghCtx?.actor,
    mutableTagOnly,
  });

  const body: Record<string, unknown> = {
    provider: inputs.provider || 'github_actions',
    sourceType: 'github_actions',
    repository: repository || undefined,
    commitSha: inputs.commitSha ?? ghCtx?.commitSha,
    branch: inputs.branch ?? ghCtx?.branch,
    workflowRunId: workflowRunId ? Number(workflowRunId) : undefined,
    serviceName: serviceName || undefined,
    environment,
    region: inputs.region || undefined,
    cluster: inputs.cluster || undefined,
    namespace: inputs.namespace || undefined,
    artifactId: inputs.artifactId || undefined,
    artifactDigest: digest,
    imageName: imageName || undefined,
    imageDigest: digest,
    containerImage: imageName && imageTag ? `${imageName}:${imageTag}` : imageName || undefined,
    version: imageTag || inputs.packageVersion || undefined,
    buildId: inputs.buildId || undefined,
    releaseGateDecisionId: inputs.releaseGateDecisionId || undefined,
    releaseGateDecision: inputs.releaseGateDecision || undefined,
    runnerId: inputs.runnerId || undefined,
    deployedBy: inputs.actor ?? ghCtx?.actor,
    externalId: inputs.externalId || undefined,
    status: mapDeploymentStatus(inputs.status),
    metadata,
  };

  return { body, warnings, digest, mutableTagOnly, idempotencyKey };
}

export function buildDeploymentGuardPayload(
  inputs: DeploymentRegistrationInputs & {
    enforcementMode?: DeploymentGuardMode;
    requireReleaseGate?: boolean;
    deploymentId?: string;
  },
  ghCtx?: GitHubWorkflowContext,
): BuildDeploymentPayloadResult {
  const base = buildDeploymentRegistrationPayload(inputs, ghCtx);
  const enforcementMode = inputs.enforcementMode ?? 'warn';
  base.body.enforcementMode = enforcementMode;
  base.body.metadata = {
    ...(base.body.metadata as Record<string, unknown>),
    enforcementMode,
  };
  if (inputs.requireReleaseGate !== undefined) {
    base.body.requireReleaseGate = inputs.requireReleaseGate;
  }
  if (inputs.deploymentId) {
    base.body.deploymentId = inputs.deploymentId;
  }
  return base;
}

export function maskSecret(value: string): void {
  void value;
  // Actions call @actions/core.setSecret locally; core stays free of @actions/core.
}
