// @bridgedai/actions-core/src/evidence-writer.ts
import * as fs from 'fs';
import * as path from 'path';
import { ensureDirForFile } from './signal-writer';

export interface AuthSessionContext {
  schemaVersion: string;
  apiUrl: string;
  organizationId: string;
  tenantId?: string;
  sessionId?: string;
  repository?: string;
  commitSha?: string;
  workflowRunId?: string;
  runAttempt?: string;
  apiReachable: boolean;
  authenticatedAt: string;
  authMethod: 'github-oidc' | 'api-key' | 'mock';
}

export interface SubjectMetadata {
  schemaVersion: string;
  subjectType: string;
  subjectId: string;
  subjectName: string;
  subjectDigest: string;
  subjectUri: string;
  metadata?: Record<string, unknown>;
}

export async function writeJsonFile(filePath: string, data: unknown): Promise<string> {
  await ensureDirForFile(filePath);
  const abs = path.resolve(filePath);
  const json = `${JSON.stringify(data, null, 2)}\n`;
  await fs.promises.writeFile(abs, json, { encoding: 'utf8', mode: 0o644 });
  return abs;
}

export async function readJsonFile<T = unknown>(filePath: string): Promise<T> {
  const abs = path.resolve(filePath);
  if (!fs.existsSync(abs)) throw new Error(`File not found: ${abs}`);
  const raw = await fs.promises.readFile(abs, 'utf8');
  return JSON.parse(raw) as T;
}

export async function writeAuthSessionContext(filePath: string, ctx: AuthSessionContext): Promise<string> {
  return writeJsonFile(filePath, ctx);
}

export async function writeSubjectMetadata(filePath: string, meta: SubjectMetadata): Promise<string> {
  return writeJsonFile(filePath, meta);
}

export function listJsonFilesInDir(dirPath: string): string[] {
  const abs = path.resolve(dirPath);
  if (!fs.existsSync(abs)) return [];
  const out: string[] = [];
  const walk = (dir: string) => {
    for (const ent of fs.readdirSync(dir, { withFileTypes: true })) {
      const p = path.join(dir, ent.name);
      if (ent.isDirectory()) walk(p);
      else if (ent.isFile() && (ent.name.endsWith('.json') || ent.name.endsWith('.jsonl'))) out.push(p);
    }
  };
  walk(abs);
  return out.sort();
}
