// @bridgedai/actions-core/src/github-attestation.ts
/**
 * Parse GitHub Artifact Attestation / actions/attest-* bundle metadata without logging raw payloads.
 */

export interface AttestationBundleSummary {
  attestationPath: string;
  attestationType: string;
  predicateType: string;
  subjectDigest: string;
  bundleVerified: boolean;
}

function digestsFromSubject(subjects: unknown): string | undefined {
  if (!Array.isArray(subjects) || subjects.length === 0) return undefined;
  const d = (subjects[0] as Record<string, unknown>)?.digest as Record<string, string> | undefined;
  const sha = String(d?.sha256 ?? '').toLowerCase();
  return sha || undefined;
}

export function parseAttestationBundle(doc: unknown): {
  bundleVerified: boolean;
  predicateType: string;
  subjectDigest?: string;
  attestationType: string;
} {
  const root = doc as Record<string, unknown>;
  const bundle = root.bundle as Record<string, unknown> | undefined;
  const bundleVerified = bundle?.verified === true || root.verified === true;

  const stmt = (root.statement ?? root.payload) as Record<string, unknown> | undefined;
  if (stmt && typeof stmt === 'object') {
    return {
      bundleVerified,
      predicateType: String(stmt.predicateType ?? ''),
      subjectDigest: digestsFromSubject(stmt.subject),
      attestationType: String(root.mediaType ?? root.attestationType ?? 'github-artestation'),
    };
  }

  const env = root.dsseEnvelope as Record<string, unknown> | undefined;
  const payloadB64 = String(env?.payload ?? '');
  if (payloadB64) {
    try {
      const stmtRaw = Buffer.from(payloadB64, 'base64').toString('utf8');
      const parsed = JSON.parse(stmtRaw) as Record<string, unknown>;
      return {
        bundleVerified,
        predicateType: String(parsed.predicateType ?? ''),
        subjectDigest: digestsFromSubject(parsed.subject),
        attestationType: 'dsse-envelope',
      };
    } catch {
      /* fall through */
    }
  }

  return {
    bundleVerified,
    predicateType: String(root.predicateType ?? ''),
    subjectDigest: undefined,
    attestationType: String(root.attestationType ?? 'unknown'),
  };
}

export function redactAttestationForSignal(doc: unknown): Record<string, unknown> {
  const parsed = parseAttestationBundle(doc);
  return {
    predicateType: parsed.predicateType,
    attestationType: parsed.attestationType,
    bundleVerified: parsed.bundleVerified,
    subjectDigest: parsed.subjectDigest ? `sha256:${parsed.subjectDigest.replace(/^sha256:/i, '')}` : undefined,
  };
}
