// @bridgedai/actions-core/src/graph-hints.ts
import type { GraphHint } from './signal-envelope';

export function graphEdge(
  fromType: string,
  fromId: string,
  toType: string,
  toId: string,
  relation: string,
): GraphHint {
  return { fromType, fromId, toType, toId, relation };
}

export function repositoryToBranch(repo: string, branch: string): GraphHint {
  return graphEdge('Repository', repo, 'Branch', `${repo}@${branch}`, 'has_branch');
}

export function branchToCommit(branch: string, commitSha: string): GraphHint {
  return graphEdge('Branch', branch, 'Commit', commitSha, 'points_to');
}

export function commitToWorkflowRun(commitSha: string, runId: string): GraphHint {
  return graphEdge('Commit', commitSha, 'WorkflowRun', runId, 'triggered');
}

export function workflowRunToRunner(runId: string, runnerName: string): GraphHint {
  return graphEdge('WorkflowRun', runId, 'Runner', runnerName, 'executed_on');
}

export function pullRequestToCommit(prNumber: number, commitSha: string): GraphHint {
  return graphEdge('PullRequest', String(prNumber), 'Commit', commitSha, 'includes');
}

export function commitToArtifact(commitSha: string, digest: string): GraphHint {
  return graphEdge('Commit', commitSha, 'Artifact', digest, 'produced');
}

export function artifactToContainerImage(artifactDigest: string, imageDigest: string): GraphHint {
  return graphEdge('Artifact', artifactDigest, 'ContainerImage', imageDigest, 'packaged_as');
}

export function containerImageToRegistry(imageDigest: string, registry: string): GraphHint {
  return graphEdge('ContainerImage', imageDigest, 'Registry', registry, 'hosted_in');
}

export function artifactToSbom(artifactDigest: string, sbomDigest: string): GraphHint {
  return graphEdge('Artifact', artifactDigest, 'SBOM', sbomDigest, 'described_by');
}

export function sbomToPackage(sbomDigest: string, packageId: string): GraphHint {
  return graphEdge('SBOM', sbomDigest, 'Package', packageId, 'lists');
}

export function packageToDependency(packageId: string, dependencyId: string): GraphHint {
  return graphEdge('Package', packageId, 'Dependency', dependencyId, 'depends_on');
}

export function artifactToDeployment(artifactDigest: string, deploymentId: string): GraphHint {
  return graphEdge('Artifact', artifactDigest, 'Deployment', deploymentId, 'deployed_as');
}

export function containerImageToDeployment(imageDigest: string, deploymentId: string): GraphHint {
  return graphEdge('ContainerImage', imageDigest, 'Deployment', deploymentId, 'deployed_as');
}

export function deploymentToEnvironment(deploymentId: string, environment: string): GraphHint {
  return graphEdge('Deployment', deploymentId, 'Environment', environment, 'targets');
}

export function deploymentToService(deploymentId: string, service: string): GraphHint {
  return graphEdge('Deployment', deploymentId, 'Service', service, 'serves');
}

export function deploymentToWorkload(deploymentId: string, workload: string): GraphHint {
  return graphEdge('Deployment', deploymentId, 'Workload', workload, 'manages');
}

export function workloadToContainerImage(workload: string, imageDigest: string): GraphHint {
  return graphEdge('Workload', workload, 'ContainerImage', imageDigest, 'runs');
}

export function artifactToAttestation(artifactDigest: string, attestationId: string): GraphHint {
  return graphEdge('Artifact', artifactDigest, 'Attestation', attestationId, 'attested_by');
}

export function attestationToBuilder(attestationId: string, builderId: string): GraphHint {
  return graphEdge('Attestation', attestationId, 'Builder', builderId, 'built_by');
}

export function attestationToSourceCommit(attestationId: string, commitSha: string): GraphHint {
  return graphEdge('Attestation', attestationId, 'SourceCommit', commitSha, 'sources_from');
}
