// @bridgedai/actions-core/src/idempotency.ts
/** Deterministic idempotency key helpers for GitHub Actions signals. */

function sanitize(part: string): string {
  return String(part ?? '')
    .trim()
    .replace(/\s+/g, '_')
    .replace(/[^a-zA-Z0-9._/@:-]/g, '-');
}

export interface WorkflowIdempotencyContext {
  ownerRepo: string;
  workflowRunId: string;
  runAttempt: string;
  jobName: string;
  workflowJobId?: string;
}

export function workflowSignalKey(
  ctx: WorkflowIdempotencyContext,
  signalType: string,
  subjectDigest?: string,
): string {
  const base = [
    'github-actions',
    sanitize(ctx.ownerRepo),
    sanitize(ctx.workflowRunId),
    sanitize(ctx.runAttempt),
    sanitize(ctx.jobName),
    sanitize(signalType),
  ].join(':');
  if (subjectDigest) {
    return `${base}:${sanitize(subjectDigest.replace(/^sha256:/i, ''))}`;
  }
  return base;
}

export function artifactSignalKey(
  ownerRepo: string,
  workflowRunId: string,
  signalType: string,
  artifactDigest: string,
): string {
  return [
    'github-actions',
    sanitize(ownerRepo),
    sanitize(workflowRunId),
    sanitize(signalType),
    sanitize(artifactDigest.replace(/^sha256:/i, '')),
  ].join(':');
}

export function ociSignalKey(registry: string, image: string, digest: string, signalType: string): string {
  return ['oci', sanitize(registry), sanitize(image), sanitize(digest.replace(/^sha256:/i, '')), sanitize(signalType)].join(
    ':',
  );
}

export function deploymentSignalKey(
  environment: string,
  service: string,
  digest: string,
  deploymentId: string,
  signalType: string,
): string {
  return [
    'deployment',
    sanitize(environment),
    sanitize(service),
    sanitize(digest.replace(/^sha256:/i, '')),
    sanitize(deploymentId),
    sanitize(signalType),
  ].join(':');
}

export function deploymentRegisteredIdempotencyKey(params: {
  repository: string;
  workflowRunId: string;
  runAttempt: string;
  environment: string;
  artifactDigest?: string;
  imageDigest?: string;
  deploymentName?: string;
}): string {
  const rawDigest = params.artifactDigest ?? params.imageDigest ?? '';
  const hex = rawDigest.replace(/^sha256:/i, '').trim();
  const digestPart = hex ? `:${hex}:` : params.deploymentName || 'none';
  return [
    'deployment-registered',
    sanitize(params.repository),
    sanitize(params.workflowRunId),
    sanitize(params.runAttempt),
    sanitize(params.environment),
    sanitize(digestPart),
  ].join(':');
}

export function runnerUsedSignalKey(ctx: WorkflowIdempotencyContext): string {
  const jobSegment = ctx.workflowJobId ? `job-${sanitize(ctx.workflowJobId)}` : sanitize(ctx.jobName);
  return [
    'github-actions',
    sanitize(ctx.ownerRepo),
    sanitize(ctx.workflowRunId),
    sanitize(ctx.runAttempt),
    jobSegment,
    'runner.used',
  ].join(':');
}
