// @bridgedai/actions-core/src/retry.ts
export async function sleep(ms: number): Promise<void> {
  await new Promise((r) => setTimeout(r, ms));
}

export interface RetryOptions {
  maxAttempts?: number;
  baseDelayMs?: number;
  isRetryable?: (error: unknown) => boolean;
}

const DEFAULT_RETRYABLE = (e: unknown): boolean => {
  if (e && typeof e === 'object' && 'status' in e) {
    const status = Number((e as { status: number }).status);
    return status === 502 || status === 503 || status === 504 || status === 429;
  }
  if (e instanceof Error && e.name === 'AbortError') return true;
  return false;
};

export async function withRetry<T>(fn: () => Promise<T>, opts: RetryOptions = {}): Promise<T> {
  const maxAttempts = opts.maxAttempts ?? 3;
  const baseDelayMs = opts.baseDelayMs ?? 250;
  const isRetryable = opts.isRetryable ?? DEFAULT_RETRYABLE;
  let lastErr: unknown;

  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (e) {
      lastErr = e;
      if (attempt >= maxAttempts - 1 || !isRetryable(e)) throw e;
      await sleep(baseDelayMs * 2 ** attempt);
    }
  }
  throw lastErr instanceof Error ? lastErr : new Error(String(lastErr));
}
