// @bridgedai/actions-core/src/runner-context.ts
import { extractGitHubContext, type GitHubWorkflowContext } from './github-context';
import { runnerUsedSignalKey, type WorkflowIdempotencyContext } from './idempotency';
import { redactUnknown } from './redaction';
import { appendSignalToFile, readSignalsFromFile } from './signal-writer';
import type { SignalEnvelope } from './signal-envelope';
import { validateSignalEnvelope } from './signal-envelope';

export { runnerUsedSignalKey as buildRunnerUsedSignalKey };

export type NormalizedRunnerContext = {
  ghCtx: GitHubWorkflowContext;
  idempotency: WorkflowIdempotencyContext;
};

/** Resolve GitHub workflow context and idempotency fields for runner.used signals. */
export function normalizeRunnerContext(opts?: {
  workflowJobId?: string;
  changedFiles?: string[];
}): NormalizedRunnerContext {
  const ghCtx = extractGitHubContext(opts?.changedFiles);
  const workflowJobId = String(opts?.workflowJobId ?? ghCtx.workflowJobId ?? '').trim() || undefined;
  if (workflowJobId) ghCtx.workflowJobId = workflowJobId;

  return {
    ghCtx,
    idempotency: {
      ownerRepo: ghCtx.repository,
      workflowRunId: ghCtx.workflowRunId,
      runAttempt: ghCtx.runAttempt,
      jobName: ghCtx.jobName,
      workflowJobId,
    },
  };
}

const REGISTER_PAYLOAD_KEYS = new Set([
  'repository',
  'workflow',
  'workflowRunId',
  'workflowJobId',
  'jobName',
  'runnerName',
  'runnerOs',
  'runnerArchitecture',
  'runnerGroup',
  'labels',
  'commitSha',
  'branch',
  'actor',
  'runAttempt',
  'buildId',
  'projectId',
  'correlationId',
  'dryRun',
]);

/** Strip unknown fields and redact sensitive values from runner registration payloads. */
export function sanitizeRunnerMetadata(payload: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(payload)) {
    if (!REGISTER_PAYLOAD_KEYS.has(key)) continue;
    if (value === undefined || value === null || value === '') continue;
    out[key] = redactUnknown(value);
  }
  return out;
}

export type BuildRunnerRegisterPayloadInput = {
  ghCtx: GitHubWorkflowContext;
  buildId?: string;
  projectId?: string;
  correlationId?: string;
  dryRun?: boolean;
  workflowJobIdOverride?: string;
};

/** Build POST /v1/runners/register body from GitHub context (backend RegisterRunnerDto fields only). */
export function buildRunnerRegisterPayload(input: BuildRunnerRegisterPayloadInput): Record<string, unknown> {
  const { ghCtx } = input;
  const workflowJobId =
    String(input.workflowJobIdOverride ?? ghCtx.workflowJobId ?? '').trim() || undefined;

  return sanitizeRunnerMetadata({
    repository: ghCtx.repository,
    workflow: ghCtx.workflowName,
    workflowRunId: ghCtx.workflowRunId,
    workflowJobId,
    jobName: ghCtx.jobName,
    runnerName: ghCtx.runnerName,
    runnerOs: ghCtx.runnerOs,
    runnerArchitecture: ghCtx.runnerArch,
    runnerGroup: ghCtx.runnerGroup,
    labels: ghCtx.runnerLabels,
    commitSha: ghCtx.commitSha,
    branch: ghCtx.branch,
    actor: ghCtx.actor,
    runAttempt: Number(ghCtx.runAttempt) || 1,
    buildId: input.buildId,
    projectId: input.projectId,
    correlationId: input.correlationId,
    dryRun: input.dryRun,
  });
}

/** Returns true when an envelope with the same idempotency key already exists in the file. */
export async function dedupeSignalEnvelope(
  filePath: string,
  signal: SignalEnvelope,
): Promise<boolean> {
  validateSignalEnvelope(signal);
  const existing = await readSignalsFromFile(filePath);
  return existing.some((s) => s.idempotencyKey === signal.idempotencyKey);
}

/**
 * Append a signal envelope; skip when the canonical idempotency key already exists.
 * Returns whether a new line was written.
 */
export async function appendSignalEnvelope(
  filePath: string,
  signal: SignalEnvelope,
): Promise<'appended' | 'deduplicated'> {
  validateSignalEnvelope(signal);
  if (await dedupeSignalEnvelope(filePath, signal)) {
    return 'deduplicated';
  }
  await appendSignalToFile(filePath, signal);
  return 'appended';
}
