// @bridgedai/actions-core/src/runner-metadata.ts
export type RunnerHostedType = 'github_hosted' | 'self_hosted' | 'unknown';

export type RunnerUsedPayload = {
  id: string;
  name?: string;
  labels: string[];
  os?: string;
  architecture?: string;
  runnerGroup?: string;
  hostedType: RunnerHostedType;
  ephemeral: boolean | 'unknown';
};

function normLabels(raw: unknown): string[] {
  if (!Array.isArray(raw)) return [];
  return raw.map((l) => String(l ?? '').trim()).filter(Boolean);
}

export function inferRunnerHostedType(input: {
  runnerName?: string;
  labels?: string[];
}): { hostedType: RunnerHostedType; selfHosted: boolean; ephemeral: boolean | 'unknown' } {
  const labels = normLabels(input.labels);
  const lowerLabels = labels.map((l) => l.toLowerCase());
  const name = (input.runnerName ?? '').toLowerCase();

  if (lowerLabels.includes('self-hosted') || name.includes('self-hosted')) {
    return { hostedType: 'self_hosted', selfHosted: true, ephemeral: 'unknown' };
  }
  if (
    name.startsWith('github actions') ||
    lowerLabels.includes('github-hosted') ||
    (name.includes('hosted') && !lowerLabels.includes('self-hosted'))
  ) {
    return { hostedType: 'github_hosted', selfHosted: false, ephemeral: true };
  }
  return { hostedType: 'unknown', selfHosted: false, ephemeral: 'unknown' };
}

export function buildRunnerUsedPayload(input: {
  runnerName?: string;
  runnerOs?: string;
  runnerArch?: string;
  runnerGroup?: string;
  labels?: string[];
  workflowJobId?: string;
}): RunnerUsedPayload & { workflowJobId?: string } {
  const labels = normLabels(input.labels);
  const inferred = inferRunnerHostedType({ runnerName: input.runnerName, labels });
  const providerId = `github:${input.runnerName ?? 'unknown'}:${input.runnerGroup ?? ''}:${[...labels].sort().join(',')}`;

  return {
    id: providerId,
    name: input.runnerName,
    labels,
    os: input.runnerOs,
    architecture: input.runnerArch,
    runnerGroup: input.runnerGroup,
    hostedType: inferred.hostedType,
    ephemeral: inferred.ephemeral,
    workflowJobId: input.workflowJobId,
  };
}

export function buildRunnerUsedNormalizedPayload(input: {
  organizationId: string;
  repository: string;
  workflowRunId: string;
  workflowJobId?: string;
  commitSha?: string;
  branch?: string;
  runner: RunnerUsedPayload;
  observedAt?: string;
}): Record<string, unknown> {
  return {
    signalType: 'runner.used',
    organizationId: input.organizationId,
    source: 'github_actions',
    repository: input.repository,
    workflowRunId: input.workflowRunId,
    workflowJobId: input.workflowJobId,
    commitSha: input.commitSha,
    branch: input.branch,
    runner: input.runner,
    observedAt: input.observedAt ?? new Date().toISOString(),
  };
}
