// @bridgedai/actions-core/src/signal-envelope.ts
import { assertVerificationLevel, type VerificationLevel } from './verification-level';

export const SIGNAL_SCHEMA_VERSION = '2026-05-01';

/** Canonical signal types emitted from GitHub Actions. */
export type SignalType =
  | 'repository.discovered'
  | 'branch.discovered'
  | 'commit.discovered'
  | 'pull_request.opened'
  | 'pull_request.merged'
  | 'pull_request.changed_files'
  | 'workflow_run.started'
  | 'workflow_run.completed'
  | 'build.started'
  | 'build.completed'
  | 'build.failed'
  | 'build.artifact_produced'
  | 'runner.used'
  | 'artifact.pushed'
  | 'artifact.tagged'
  | 'container_image.pushed'
  | 'container_image.scanned'
  | 'container_image.deployed'
  | 'container_layer.discovered'
  | 'sbom.generated'
  | 'package.discovered'
  | 'dependency.discovered'
  | 'vulnerability.detected'
  | 'vulnerability.fixed'
  | 'vex.statement'
  | 'signature.observed'
  | 'signature.verified'
  | 'attestation.observed'
  | 'attestation.verified'
  | 'slsa.provenance.observed'
  | 'slsa.provenance.verified'
  | 'intoto.statement.observed'
  | 'intoto.statement.verified'
  | 'policy.evaluated'
  | 'deployment.started'
  | 'deployment.completed'
  | 'deployment.failed'
  | 'workload.updated';

export interface SignalSubject {
  repository?: string;
  commitSha?: string;
  workflowRunId?: string;
  runAttempt?: string;
  jobName?: string;
  artifactDigest?: string;
  imageDigest?: string;
  environment?: string;
  service?: string;
}

export interface GraphHint {
  fromType: string;
  fromId: string;
  toType: string;
  toId: string;
  relation: string;
}

export interface SignalEnvelope {
  organizationId: string;
  provider: 'github';
  sourceSystem: 'github-actions';
  signalType: SignalType | string;
  schemaVersion: string;
  externalId: string;
  idempotencyKey: string;
  observedAt: string;
  confidence: 'high' | 'medium' | 'low';
  verificationLevel: VerificationLevel;
  subject: SignalSubject;
  rawPayload: Record<string, unknown>;
  normalizedPayload: Record<string, unknown>;
  evidenceHints: string[];
  graphHints: GraphHint[];
  metadata: Record<string, unknown>;
}

export interface CreateSignalOptions {
  organizationId: string;
  signalType: SignalType | string;
  idempotencyKey: string;
  externalId?: string;
  verificationLevel?: VerificationLevel;
  confidence?: 'high' | 'medium' | 'low';
  subject?: SignalSubject;
  rawPayload?: Record<string, unknown>;
  normalizedPayload?: Record<string, unknown>;
  evidenceHints?: string[];
  graphHints?: GraphHint[];
  metadata?: Record<string, unknown>;
  observedAt?: string;
}

export function createSignalEnvelope(opts: CreateSignalOptions): SignalEnvelope {
  const orgId = String(opts.organizationId ?? '').trim();
  if (!orgId) throw new Error('organizationId is required for signal envelope');

  const signalType = String(opts.signalType ?? '').trim();
  if (!signalType) throw new Error('signalType is required for signal envelope');

  const idempotencyKey = String(opts.idempotencyKey ?? '').trim();
  if (!idempotencyKey) throw new Error('idempotencyKey is required for signal envelope');

  const verificationLevel = opts.verificationLevel ?? 'provider_authoritative';
  assertVerificationLevel(verificationLevel);

  return {
    organizationId: orgId,
    provider: 'github',
    sourceSystem: 'github-actions',
    signalType,
    schemaVersion: SIGNAL_SCHEMA_VERSION,
    externalId: opts.externalId ?? idempotencyKey,
    idempotencyKey,
    observedAt: opts.observedAt ?? new Date().toISOString(),
    confidence: opts.confidence ?? 'high',
    verificationLevel,
    subject: opts.subject ?? {},
    rawPayload: opts.rawPayload ?? {},
    normalizedPayload: opts.normalizedPayload ?? {},
    evidenceHints: opts.evidenceHints ?? [],
    graphHints: opts.graphHints ?? [],
    metadata: opts.metadata ?? {},
  };
}

export function validateSignalEnvelope(envelope: unknown): SignalEnvelope {
  if (!envelope || typeof envelope !== 'object') {
    throw new Error('Signal envelope must be an object');
  }
  const e = envelope as Record<string, unknown>;

  const required = [
    'organizationId',
    'provider',
    'sourceSystem',
    'signalType',
    'schemaVersion',
    'externalId',
    'idempotencyKey',
    'observedAt',
    'confidence',
    'verificationLevel',
  ] as const;

  for (const key of required) {
    if (typeof e[key] !== 'string' || !String(e[key]).trim()) {
      throw new Error(`Signal envelope missing or invalid field: ${key}`);
    }
  }

  if (e.provider !== 'github') throw new Error('Signal envelope provider must be github');
  if (e.sourceSystem !== 'github-actions') {
    throw new Error('Signal envelope sourceSystem must be github-actions');
  }

  assertVerificationLevel(String(e.verificationLevel));

  if (!e.subject || typeof e.subject !== 'object') {
    throw new Error('Signal envelope subject must be an object');
  }
  if (!Array.isArray(e.evidenceHints)) throw new Error('Signal envelope evidenceHints must be an array');
  if (!Array.isArray(e.graphHints)) throw new Error('Signal envelope graphHints must be an array');
  if (!e.rawPayload || typeof e.rawPayload !== 'object') {
    throw new Error('Signal envelope rawPayload must be an object');
  }
  if (!e.normalizedPayload || typeof e.normalizedPayload !== 'object') {
    throw new Error('Signal envelope normalizedPayload must be an object');
  }

  return e as unknown as SignalEnvelope;
}
