// @bridgedai/actions-core/src/signal-writer.ts
import * as fs from 'fs';
import * as path from 'path';
import { validateSignalEnvelope, type SignalEnvelope } from './signal-envelope';

export const DEFAULT_SIGNALS_FILE = '.bridgedai/signals.jsonl';
export const DEFAULT_SIGNALS_BATCH_FILE = '.bridgedai/signals-batch.json';

export async function ensureDirForFile(filePath: string): Promise<void> {
  const dir = path.dirname(path.resolve(filePath));
  await fs.promises.mkdir(dir, { recursive: true });
}

export async function readSignalsFromFile(filePath: string): Promise<SignalEnvelope[]> {
  const abs = path.resolve(filePath);
  if (!fs.existsSync(abs)) return [];
  const raw = await fs.promises.readFile(abs, 'utf8');
  const trimmed = raw.trim();
  if (!trimmed) return [];

  if (abs.endsWith('.jsonl')) {
    const lines = trimmed.split('\n').filter((l) => l.trim());
    return lines.map((line, i) => {
      try {
        return validateSignalEnvelope(JSON.parse(line) as unknown);
      } catch (e) {
        throw new Error(`Invalid signal at ${abs}:${i + 1}: ${e instanceof Error ? e.message : String(e)}`);
      }
    });
  }

  const parsed = JSON.parse(trimmed) as unknown;
  if (Array.isArray(parsed)) {
    return parsed.map((item, i) => {
      try {
        return validateSignalEnvelope(item);
      } catch (e) {
        throw new Error(`Invalid signal at ${abs}[${i}]: ${e instanceof Error ? e.message : String(e)}`);
      }
    });
  }
  if (parsed && typeof parsed === 'object' && Array.isArray((parsed as Record<string, unknown>).signals)) {
    return ((parsed as Record<string, unknown>).signals as unknown[]).map((item, i) => {
      try {
        return validateSignalEnvelope(item);
      } catch (e) {
        throw new Error(`Invalid signal in batch at ${abs}[${i}]: ${e instanceof Error ? e.message : String(e)}`);
      }
    });
  }
  return [validateSignalEnvelope(parsed)];
}

export async function appendSignalToFile(filePath: string, signal: SignalEnvelope): Promise<void> {
  validateSignalEnvelope(signal);
  await ensureDirForFile(filePath);
  const abs = path.resolve(filePath);
  const line = `${JSON.stringify(signal)}\n`;
  await fs.promises.appendFile(abs, line, { encoding: 'utf8', mode: 0o644 });
}

export async function writeSignalsBatchFile(
  filePath: string,
  signals: SignalEnvelope[],
  metadata?: Record<string, unknown>,
): Promise<void> {
  for (const s of signals) validateSignalEnvelope(s);
  await ensureDirForFile(filePath);
  const abs = path.resolve(filePath);
  const doc = {
    schemaVersion: '2026-05-01',
    type: 'bridgedai_signal_batch',
    signalCount: signals.length,
    signals,
    metadata: metadata ?? {},
  };
  await fs.promises.writeFile(abs, `${JSON.stringify(doc, null, 2)}\n`, { encoding: 'utf8', mode: 0o644 });
}

export async function mergeSignalsToBatch(
  outputFile: string,
  newSignals: SignalEnvelope[],
): Promise<SignalEnvelope[]> {
  const existing = fs.existsSync(path.resolve(outputFile)) ? await readSignalsFromFile(outputFile) : [];
  const byKey = new Map<string, SignalEnvelope>();
  for (const s of existing) byKey.set(s.idempotencyKey, s);
  for (const s of newSignals) byKey.set(s.idempotencyKey, s);
  const merged = [...byKey.values()];
  await writeSignalsBatchFile(outputFile, merged);
  return merged;
}

export function dedupeSignalsByIdempotencyKey(signals: SignalEnvelope[]): SignalEnvelope[] {
  const byKey = new Map<string, SignalEnvelope>();
  for (const s of signals) {
    validateSignalEnvelope(s);
    byKey.set(s.idempotencyKey, s);
  }
  return [...byKey.values()];
}
