// @bridgedai/actions-core/src/validation-url.ts
/** Normalize API base URL (no trailing slash). */
export function normalizeBaseUrl(raw: string, fieldName = 'api-base'): string {
  const trimmed = String(raw ?? '').trim();
  if (!trimmed) throw new Error(`Missing ${fieldName}`);
  let u: URL;
  try {
    u = new URL(trimmed);
  } catch {
    throw new Error(`Invalid ${fieldName}: ${trimmed}`);
  }
  const host = u.hostname.toLowerCase();
  const isLocalHttp =
    host === 'localhost' || host === '127.0.0.1' || host === '::1' || host === '[::1]';
  if (u.protocol !== 'https:' && !isLocalHttp) {
    throw new Error(`${fieldName} must use https:// unless targeting localhost`);
  }
  return u.toString().replace(/\/+$/, '');
}

/** Matches @actions/core getInput and GitHub runner env names (hyphens preserved). */
function githubInputEnvKey(name: string): string {
  return `INPUT_${name.replace(/ /g, '_').toUpperCase()}`;
}

/** Legacy test/E2E form that mapped hyphens to underscores. */
function legacyInputEnvKey(name: string): string {
  return `INPUT_${name.replace(/\s+/g, '_').replace(/-/g, '_').toUpperCase()}`;
}

/**
 * Read a GitHub Actions input from runner env (hyphenated names, e.g. api-key → INPUT_API-KEY).
 * Prefer getOptionalInput/getRequiredInput in action code — they call @actions/core getInput first.
 */
export function readWorkflowInput(name: string): string {
  const githubKey = githubInputEnvKey(name);
  const legacyKey = legacyInputEnvKey(name);
  const raw = process.env[githubKey] ?? (githubKey !== legacyKey ? process.env[legacyKey] : undefined);
  return String(raw ?? '').trim();
}

/** Input value with optional process.env fallbacks (e.g. BRIDGED_API_KEY). */
export function resolveWorkflowInput(name: string, ...envNames: string[]): string {
  const fromInput = readWorkflowInput(name);
  if (fromInput) return fromInput;
  for (const envName of envNames) {
    const v = String(process.env[envName] ?? '').trim();
    if (v) return v;
  }
  return '';
}

/** Required input with env fallbacks; throws if all are empty. */
export function requireWorkflowInput(name: string, ...envNames: string[]): string {
  const v = resolveWorkflowInput(name, ...envNames);
  if (v) return v;
  const envHint = envNames.length ? ` or ${envNames.join(' / ')} env` : '';
  throw new Error(`Missing required ${name} input${envHint}`);
}
