// @bridgedai/actions-core/dist/auth-credentials.js
"use strict";
/**
 * Resolve BridgedAI credentials for HTTP clients.
 * Org API keys (brdg_*) must use x-api-key, not Authorization: Bearer.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ORG_ID_PATTERN = exports.MAX_BRIDGED_CREDENTIAL_LENGTH = exports.BRIDGED_API_KEY_PATTERN = void 0;
exports.isBridgedApiKey = isBridgedApiKey;
exports.isValidBridgedOrgId = isValidBridgedOrgId;
exports.splitBridgedCredential = splitBridgedCredential;
exports.resolveBridgedAuth = resolveBridgedAuth;
exports.requireBridgedAuth = requireBridgedAuth;
/** Must match backend ApiKeyService.BRIDGED_API_KEY_PATTERN */
exports.BRIDGED_API_KEY_PATTERN = /^brdg_(live|test)_sk_(ak_[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})_([A-Za-z0-9_-]{32,64})$/i;
exports.MAX_BRIDGED_CREDENTIAL_LENGTH = 256;
exports.ORG_ID_PATTERN = /^org_[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
function isBridgedApiKey(value) {
    const t = String(value ?? '').trim();
    return t.length > 0 && t.length <= exports.MAX_BRIDGED_CREDENTIAL_LENGTH && exports.BRIDGED_API_KEY_PATTERN.test(t);
}
function isValidBridgedOrgId(orgId) {
    return exports.ORG_ID_PATTERN.test(String(orgId ?? '').trim());
}
/** Split a single secret into api-key vs JWT Bearer. */
function splitBridgedCredential(token) {
    const t = String(token ?? '').trim();
    if (!t)
        return {};
    if (t.length > exports.MAX_BRIDGED_CREDENTIAL_LENGTH) {
        throw new Error('BridgedAI credential exceeds maximum allowed length');
    }
    if (isBridgedApiKey(t))
        return { apiKey: t };
    return { accessToken: t };
}
/**
 * Prefer explicit api-key; if access-token is a brdg_* key, route to x-api-key even when both are set.
 */
function resolveBridgedAuth(input) {
    const apiKeyRaw = String(input.apiKey ?? '').trim();
    const tokenRaw = String(input.accessToken ?? input.bridgedaiToken ?? '').trim();
    if (apiKeyRaw.length > exports.MAX_BRIDGED_CREDENTIAL_LENGTH || tokenRaw.length > exports.MAX_BRIDGED_CREDENTIAL_LENGTH) {
        throw new Error('BridgedAI credential exceeds maximum allowed length');
    }
    if (apiKeyRaw) {
        if (isBridgedApiKey(apiKeyRaw))
            return { apiKey: apiKeyRaw };
        return { accessToken: apiKeyRaw };
    }
    if (tokenRaw)
        return splitBridgedCredential(tokenRaw);
    return {};
}
function requireBridgedAuth(input, message) {
    const cred = resolveBridgedAuth(input);
    if (!cred.apiKey && !cred.accessToken) {
        throw new Error(message ?? 'Missing BridgedAI credentials (api-key or access-token required)');
    }
    return cred;
}
