// @bridgedai/actions-core/dist/bridgedai-client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.uploadGenericEvidence = exports.uploadAttestation = exports.uploadSbom = exports.BridgedHttpClient = exports.BridgedApiError = exports.requireWorkflowInput = exports.resolveWorkflowInput = exports.readWorkflowInput = exports.normalizeBaseUrl = void 0;
exports.evaluateReleaseGate = evaluateReleaseGate;
exports.evaluateDeploymentGuard = evaluateDeploymentGuard;
exports.registerDeployment = registerDeployment;
exports.ingestSbom = ingestSbom;
exports.ingestAttestation = ingestAttestation;
exports.getCapabilities = getCapabilities;
exports.resolveBuild = resolveBuild;
exports.resolveBuildForCi = resolveBuildForCi;
exports.ensureBuild = ensureBuild;
exports.extractBuildIdFromResolveResponse = extractBuildIdFromResolveResponse;
exports.ingestEvidence = ingestEvidence;
exports.verifyArtifactSupplyChain = verifyArtifactSupplyChain;
exports.getBuildSupplyChainVerification = getBuildSupplyChainVerification;
exports.getCurrentUserMe = getCurrentUserMe;
exports.sha256HexOfBuffer = sha256HexOfBuffer;
exports.sha256HexOfFile = sha256HexOfFile;
exports.listFilesUnderDirectory = listFilesUnderDirectory;
exports.computeDirectoryManifestSha256 = computeDirectoryManifestSha256;
const crypto_1 = require("crypto");
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const auth_credentials_1 = require("./auth-credentials");
const validation_url_1 = require("./validation-url");
const redaction_1 = require("./redaction");
const retry_1 = require("./retry");
var validation_url_2 = require("./validation-url");
Object.defineProperty(exports, "normalizeBaseUrl", { enumerable: true, get: function () { return validation_url_2.normalizeBaseUrl; } });
Object.defineProperty(exports, "readWorkflowInput", { enumerable: true, get: function () { return validation_url_2.readWorkflowInput; } });
Object.defineProperty(exports, "resolveWorkflowInput", { enumerable: true, get: function () { return validation_url_2.resolveWorkflowInput; } });
Object.defineProperty(exports, "requireWorkflowInput", { enumerable: true, get: function () { return validation_url_2.requireWorkflowInput; } });
class BridgedApiError extends Error {
    status;
    code;
    detail;
    bodySnippet;
    constructor(opts) {
        super(opts.message);
        this.name = 'BridgedApiError';
        this.status = opts.status;
        this.code = opts.code;
        this.detail = opts.detail;
        this.bodySnippet = opts.bodySnippet ?? '';
    }
}
exports.BridgedApiError = BridgedApiError;
function problemMessage(status, body, rawText) {
    if (body && typeof body === 'object') {
        const o = body;
        const detail = typeof o.detail === 'string' ? o.detail : typeof o.message === 'string' ? o.message : undefined;
        const title = typeof o.title === 'string' ? o.title : undefined;
        const code = typeof o.code === 'string' ? o.code : typeof o.error === 'string' ? o.error : undefined;
        const msg = [title, detail].filter(Boolean).join(': ') || `HTTP ${status}`;
        return { msg, code, detail };
    }
    const t = rawText.length > 2000 ? `${rawText.slice(0, 2000)}…` : rawText;
    return { msg: t || `HTTP ${status}` };
}
class BridgedHttpClient {
    baseUrl;
    apiKey;
    accessToken;
    orgId;
    timeoutMs;
    preferAuthorizationApiKey;
    constructor(opts) {
        this.baseUrl = (0, validation_url_1.normalizeBaseUrl)(opts.baseUrl);
        let apiKey = String(opts.apiKey ?? '').trim();
        let accessToken = String(opts.accessToken ?? '').trim();
        if (apiKey.length > auth_credentials_1.MAX_BRIDGED_CREDENTIAL_LENGTH || accessToken.length > auth_credentials_1.MAX_BRIDGED_CREDENTIAL_LENGTH) {
            throw new Error('BridgedAI credential exceeds maximum allowed length');
        }
        if (accessToken && (0, auth_credentials_1.isBridgedApiKey)(accessToken)) {
            apiKey = accessToken;
            accessToken = '';
        }
        this.apiKey = apiKey;
        this.accessToken = accessToken;
        this.orgId = String(opts.orgId ?? '').trim();
        this.timeoutMs = opts.timeoutMs ?? 30_000;
        this.preferAuthorizationApiKey = opts.preferAuthorizationApiKey ?? false;
        if (!this.apiKey && !this.accessToken) {
            throw new Error('Missing BridgedAI credentials (apiKey or accessToken required)');
        }
        if (!this.orgId)
            throw new Error('Missing BridgedAI organization id');
        if (!(0, auth_credentials_1.isValidBridgedOrgId)(this.orgId)) {
            throw new Error('Invalid BridgedAI organization id format (expected org_<uuid>)');
        }
    }
    /** Credential used for log redaction (never logged itself). */
    redactionSecret() {
        return this.accessToken || this.apiKey;
    }
    authHeaders() {
        if (this.accessToken) {
            return {
                Authorization: `Bearer ${this.accessToken}`,
                'x-org-id': this.orgId,
            };
        }
        if (this.preferAuthorizationApiKey) {
            return {
                Authorization: `ApiKey ${this.apiKey}`,
                'x-org-id': this.orgId,
            };
        }
        return {
            'x-api-key': this.apiKey,
            'x-org-id': this.orgId,
        };
    }
    async requestJson(opts) {
        const url = `${this.baseUrl}${opts.path.startsWith('/') ? opts.path : `/${opts.path}`}`;
        const allowRetry = opts.allowRetry ?? opts.method === 'GET';
        return (0, retry_1.withRetry)(async () => {
            const controller = new AbortController();
            const timer = setTimeout(() => controller.abort(), this.timeoutMs);
            try {
                const headers = {
                    accept: 'application/json',
                    ...this.authHeaders(),
                };
                const init = { method: opts.method, headers, signal: controller.signal };
                if (opts.body !== undefined && opts.method === 'POST') {
                    headers['content-type'] = 'application/json';
                    init.body = JSON.stringify(opts.body);
                }
                const res = await fetch(url, init);
                const text = await res.text();
                let parsed = undefined;
                if (text) {
                    try {
                        parsed = JSON.parse(text);
                    }
                    catch {
                        parsed = { raw: text };
                    }
                }
                if (res.ok)
                    return parsed;
                const { msg, code, detail } = problemMessage(res.status, parsed, text);
                const secret = this.redactionSecret();
                const snippet = (0, redaction_1.redactSecrets)(text.slice(0, 800), secret);
                let userMsg = (0, redaction_1.redactSecrets)(msg, secret);
                if (res.status === 422) {
                    userMsg += ' (Unprocessable entity — verify JSON body matches the public CI API contract.)';
                }
                const retryable = res.status === 502 || res.status === 503 || res.status === 504;
                if (allowRetry && retryable) {
                    throw Object.assign(new BridgedApiError({
                        status: res.status,
                        message: `BridgedAI API ${res.status}: ${userMsg}`,
                        code,
                        detail,
                        bodySnippet: snippet,
                    }), { retryable: true });
                }
                throw new BridgedApiError({
                    status: res.status,
                    message: `BridgedAI API ${res.status}: ${userMsg}`,
                    code,
                    detail,
                    bodySnippet: snippet,
                });
            }
            catch (e) {
                if (e instanceof BridgedApiError)
                    throw e;
                const aborted = e instanceof Error && e.name === 'AbortError';
                if (aborted) {
                    throw new BridgedApiError({
                        status: 0,
                        message: `BridgedAI request timed out after ${this.timeoutMs}ms (${opts.method} ${opts.path})`,
                    });
                }
                throw e;
            }
            finally {
                clearTimeout(timer);
            }
        }, {
            maxAttempts: allowRetry ? 3 : 1,
            isRetryable: (e) => {
                if (e && typeof e === 'object' && 'retryable' in e)
                    return !!e.retryable;
                if (e instanceof BridgedApiError)
                    return e.status === 502 || e.status === 503 || e.status === 504;
                return false;
            },
        });
    }
}
exports.BridgedHttpClient = BridgedHttpClient;
async function evaluateReleaseGate(client, body) {
    return client.requestJson({
        method: 'POST',
        path: '/v1/enforcement/release-gate/evaluate',
        body,
        allowRetry: false,
    });
}
async function evaluateDeploymentGuard(client, body) {
    return client.requestJson({
        method: 'POST',
        path: '/v1/deployment-guard/evaluate',
        body,
        allowRetry: false,
    });
}
async function registerDeployment(client, body) {
    return client.requestJson({
        method: 'POST',
        path: '/v1/deployments/register',
        body,
        allowRetry: false,
    });
}
async function ingestSbom(client, body) {
    return client.requestJson({
        method: 'POST',
        path: '/v1/ingest/sbom',
        body,
        allowRetry: false,
    });
}
async function ingestAttestation(client, body) {
    return client.requestJson({
        method: 'POST',
        path: '/v1/ingest/attestation',
        body,
        allowRetry: false,
    });
}
exports.uploadSbom = ingestSbom;
exports.uploadAttestation = ingestAttestation;
async function getCapabilities(_client) {
    void _client;
    const { getActionsCapabilities } = await Promise.resolve().then(() => __importStar(require('./capabilities')));
    return getActionsCapabilities();
}
async function resolveBuild(client, body) {
    return client.requestJson({
        method: 'POST',
        path: '/v1/builds/resolve',
        body,
        allowRetry: false,
    });
}
/** CI resolve: always send workflowRunId and commitSha so the API can match legacy builds. */
async function resolveBuildForCi(client, params) {
    return resolveBuild(client, {
        orgId: params.orgId,
        repoFullName: params.repoFullName,
        workflowRunId: params.workflowRunId,
        commitSha: params.commitSha,
        branch: params.branch,
    });
}
/** Idempotent CI registration — creates build immediately without waiting for webhooks. */
async function ensureBuild(client, body) {
    return client.requestJson({
        method: 'POST',
        path: '/v1/builds/ensure',
        body,
        allowRetry: false,
    });
}
function extractBuildIdFromResolveResponse(res) {
    if (!res || typeof res !== 'object')
        return '';
    const o = res;
    if (typeof o.buildId === 'string' && o.buildId)
        return o.buildId;
    if (typeof o.build_id === 'string' && o.build_id)
        return o.build_id;
    const b = o.build;
    if (b && typeof b === 'object') {
        const id = b.id;
        if (typeof id === 'string' && id)
            return id;
    }
    return '';
}
async function ingestEvidence(client, body) {
    return client.requestJson({
        method: 'POST',
        path: '/v1/ingest/evidence',
        body,
        allowRetry: false,
    });
}
exports.uploadGenericEvidence = ingestEvidence;
async function verifyArtifactSupplyChain(client, artifactId, body) {
    const urlPath = `/v1/artifacts/${encodeURIComponent(artifactId)}/verify-supply-chain`;
    return client.requestJson({
        method: 'POST',
        path: urlPath,
        body,
        allowRetry: false,
    });
}
async function getBuildSupplyChainVerification(client, buildId) {
    const urlPath = `/v1/builds/${encodeURIComponent(buildId)}/supply-chain-verification`;
    return client.requestJson({
        method: 'GET',
        path: urlPath,
        allowRetry: true,
    });
}
async function getCurrentUserMe(client) {
    return client.requestJson({
        method: 'GET',
        path: '/v1/users/me',
        allowRetry: true,
    });
}
function sha256HexOfBuffer(buf) {
    return (0, crypto_1.createHash)('sha256').update(buf).digest('hex');
}
function sha256HexOfFile(filePath, fsMod = fs) {
    return sha256HexOfBuffer(fsMod.readFileSync(filePath));
}
function listFilesUnderDirectory(rootAbs, fsMod = fs) {
    const root = path.resolve(rootAbs);
    const files = [];
    const walk = (dir) => {
        for (const ent of fsMod.readdirSync(dir, { withFileTypes: true })) {
            if (ent.name === '.git' && ent.isDirectory())
                continue;
            const abs = path.join(dir, ent.name);
            if (ent.isDirectory())
                walk(abs);
            else if (ent.isFile())
                files.push(abs);
        }
    };
    walk(root);
    return files.sort((a, b) => a.localeCompare(b));
}
function computeDirectoryManifestSha256(rootAbs, fsMod = fs) {
    const root = path.resolve(rootAbs);
    const h = (0, crypto_1.createHash)('sha256');
    const files = listFilesUnderDirectory(root, fsMod);
    for (const abs of files) {
        const rel = path.relative(root, abs).split(path.sep).join('/');
        const fileHash = sha256HexOfFile(abs, fsMod);
        h.update(rel);
        h.update('\n');
        h.update(fileHash);
        h.update('\n');
    }
    return h.digest('hex');
}
