// @bridgedai/actions-core/dist/capabilities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BRIDGED_ACTIONS_CAPABILITIES = void 0;
exports.getActionsCapabilities = getActionsCapabilities;
exports.isDeploymentRegistrationSupported = isDeploymentRegistrationSupported;
exports.isSigstoreCryptoSupported = isSigstoreCryptoSupported;
/**
 * Static CI capabilities contract for BridgedAI GitHub Actions.
 * Previously served by GET /v1/actions/capabilities on the backend; now owned by the actions ecosystem.
 * @see private/actions-dev/docs/API_COMPATIBILITY.md
 */
exports.BRIDGED_ACTIONS_CAPABILITIES = {
    schemaVersion: '2026-05-01',
    releaseGate: {
        supported: true,
        endpoint: '/v1/enforcement/release-gate/evaluate',
        buildResolution: ['buildId', 'repoFullName+workflowRunId', 'repoFullName+commitSha'],
        requiredPermission: 'release-gates:evaluate',
    },
    buildResolution: {
        supported: true,
        endpoint: '/v1/builds/resolve',
        requiredPermission: 'release-gates:evaluate',
    },
    ingestion: {
        sbom: true,
        attestation: true,
        genericEvidence: true,
        sbomLegacyBuildIdOptional: true,
        sbomEvidenceEmissionRequiresBuildId: true,
        sbomCiRequiresBuildIdWhen: ['source=github-actions', 'linkEvidence=true'],
        supportedKinds: [
            'sbom',
            'intoto_statement',
            'slsa_provenance',
            'sigstore_bundle',
            'sigstore_signature',
            'rekor_entry',
            'artifact_metadata',
        ],
        endpoints: {
            sbom: '/v1/ingest/sbom',
            attestation: '/v1/ingest/attestation',
            evidence: '/v1/ingest/evidence',
        },
        uploadPermissions: ['evidence:upload', 'projects:update'],
    },
    verification: {
        supplyChainStructural: true,
        sigstoreCrypto: false,
        rekorOnline: false,
        artifactVerifyEndpoint: '/v1/artifacts/:artifactId/verify-supply-chain',
        buildVerificationEndpoint: '/v1/builds/:buildId/supply-chain-verification',
    },
    deploymentRegistration: {
        supported: false,
        dryRunOnly: true,
    },
    signals: {
        /** Native POST /v1/signals/ingest when backend supports it; actions client falls back to evidence bridge. */
        nativeIngestEndpoint: true,
        nativeIngestPath: '/v1/signals/ingest',
        ingestViaEvidence: true,
        dedicatedIngestEndpoint: true,
        batchFormat: 'bridgedai-signal-batch-v1',
    },
};
/** @deprecated Use getActionsCapabilities() — backend GET /v1/actions/capabilities removed. */
function getActionsCapabilities() {
    return exports.BRIDGED_ACTIONS_CAPABILITIES;
}
function isDeploymentRegistrationSupported(cap = exports.BRIDGED_ACTIONS_CAPABILITIES) {
    const dep = cap.deploymentRegistration;
    return dep?.supported === true;
}
function isSigstoreCryptoSupported(cap = exports.BRIDGED_ACTIONS_CAPABILITIES) {
    const ver = cap.verification;
    return ver?.sigstoreCrypto === true;
}
