// @bridgedai/actions-core/dist/files.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeDirectoryManifestSha256 = exports.listFilesUnderDirectory = exports.sha256HexOfFile = exports.sha256HexOfBuffer = void 0;
exports.readTextFile = readTextFile;
exports.fileExists = fileExists;
exports.sha256HexOfString = sha256HexOfString;
const crypto_1 = require("crypto");
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
var files_hash_1 = require("./files-hash");
Object.defineProperty(exports, "sha256HexOfBuffer", { enumerable: true, get: function () { return files_hash_1.sha256HexOfBuffer; } });
Object.defineProperty(exports, "sha256HexOfFile", { enumerable: true, get: function () { return files_hash_1.sha256HexOfFile; } });
Object.defineProperty(exports, "listFilesUnderDirectory", { enumerable: true, get: function () { return files_hash_1.listFilesUnderDirectory; } });
Object.defineProperty(exports, "computeDirectoryManifestSha256", { enumerable: true, get: function () { return files_hash_1.computeDirectoryManifestSha256; } });
function readTextFile(filePath) {
    const abs = path.resolve(filePath);
    if (!fs.existsSync(abs))
        throw new Error(`File not found: ${abs}`);
    return fs.readFileSync(abs, 'utf8');
}
function fileExists(filePath) {
    return fs.existsSync(path.resolve(filePath));
}
function sha256HexOfString(text) {
    return (0, crypto_1.createHash)('sha256').update(text, 'utf8').digest('hex');
}
