// @bridgedai/actions-core/dist/github-attestation.js
"use strict";
/**
 * Parse GitHub Artifact Attestation / actions/attest-* bundle metadata without logging raw payloads.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseAttestationBundle = parseAttestationBundle;
exports.redactAttestationForSignal = redactAttestationForSignal;
function digestsFromSubject(subjects) {
    if (!Array.isArray(subjects) || subjects.length === 0)
        return undefined;
    const d = subjects[0]?.digest;
    const sha = String(d?.sha256 ?? '').toLowerCase();
    return sha || undefined;
}
function parseAttestationBundle(doc) {
    const root = doc;
    const bundle = root.bundle;
    const bundleVerified = bundle?.verified === true || root.verified === true;
    const stmt = (root.statement ?? root.payload);
    if (stmt && typeof stmt === 'object') {
        return {
            bundleVerified,
            predicateType: String(stmt.predicateType ?? ''),
            subjectDigest: digestsFromSubject(stmt.subject),
            attestationType: String(root.mediaType ?? root.attestationType ?? 'github-artestation'),
        };
    }
    const env = root.dsseEnvelope;
    const payloadB64 = String(env?.payload ?? '');
    if (payloadB64) {
        try {
            const stmtRaw = Buffer.from(payloadB64, 'base64').toString('utf8');
            const parsed = JSON.parse(stmtRaw);
            return {
                bundleVerified,
                predicateType: String(parsed.predicateType ?? ''),
                subjectDigest: digestsFromSubject(parsed.subject),
                attestationType: 'dsse-envelope',
            };
        }
        catch {
            /* fall through */
        }
    }
    return {
        bundleVerified,
        predicateType: String(root.predicateType ?? ''),
        subjectDigest: undefined,
        attestationType: String(root.attestationType ?? 'unknown'),
    };
}
function redactAttestationForSignal(doc) {
    const parsed = parseAttestationBundle(doc);
    return {
        predicateType: parsed.predicateType,
        attestationType: parsed.attestationType,
        bundleVerified: parsed.bundleVerified,
        subjectDigest: parsed.subjectDigest ? `sha256:${parsed.subjectDigest.replace(/^sha256:/i, '')}` : undefined,
    };
}
