// @bridgedai/actions-core/dist/github-context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractGitHubContext = extractGitHubContext;
exports.githubContextToSignalSubject = githubContextToSignalSubject;
exports.safeGithubEnvRecord = safeGithubEnvRecord;
function env(name) {
    return String(process.env[name] ?? '').trim();
}
function parsePullRequestNumber(eventName, eventPath) {
    if (eventName !== 'pull_request' && eventName !== 'pull_request_target')
        return undefined;
    if (!eventPath)
        return undefined;
    try {
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        const fs = require('fs');
        const raw = fs.readFileSync(eventPath, 'utf8');
        const event = JSON.parse(raw);
        const pr = event.pull_request;
        const num = pr?.number;
        return typeof num === 'number' ? num : undefined;
    }
    catch {
        return undefined;
    }
}
function extractGitHubContext(changedFiles) {
    const repository = env('GITHUB_REPOSITORY');
    const commitSha = env('GITHUB_SHA');
    const workflowRunId = env('GITHUB_RUN_ID');
    const runAttempt = env('GITHUB_RUN_ATTEMPT') || '1';
    const jobName = env('GITHUB_JOB') || 'unknown';
    if (!repository)
        throw new Error('GITHUB_REPOSITORY is required for GitHub context');
    if (!commitSha)
        throw new Error('GITHUB_SHA is required for GitHub context');
    if (!workflowRunId)
        throw new Error('GITHUB_RUN_ID is required for GitHub context');
    const eventName = env('GITHUB_EVENT_NAME') || undefined;
    const eventPath = env('GITHUB_EVENT_PATH') || undefined;
    return {
        repository,
        repositoryId: env('GITHUB_REPOSITORY_ID') || undefined,
        repositoryOwner: env('GITHUB_REPOSITORY_OWNER') || undefined,
        branch: env('GITHUB_REF_NAME') || undefined,
        ref: env('GITHUB_REF') || undefined,
        refType: env('GITHUB_REF_TYPE') || undefined,
        commitSha,
        actor: env('GITHUB_ACTOR') || undefined,
        eventName,
        workflowName: env('GITHUB_WORKFLOW') || undefined,
        workflowRef: env('GITHUB_WORKFLOW_REF') || undefined,
        workflowRunId,
        workflowRunNumber: env('GITHUB_RUN_NUMBER') || undefined,
        runAttempt,
        jobName,
        runnerName: env('RUNNER_NAME') || undefined,
        runnerOs: env('RUNNER_OS') || undefined,
        runnerArch: env('RUNNER_ARCH') || undefined,
        runnerLabels: inferRunnerLabelsFromEnv(),
        runnerGroup: env('RUNNER_GROUP') || undefined,
        workflowJobId: env('GITHUB_JOB_ID') || undefined,
        pullRequestNumber: parsePullRequestNumber(eventName ?? '', eventPath),
        environment: env('GITHUB_ENVIRONMENT') || undefined,
        changedFiles,
    };
}
function inferRunnerLabelsFromEnv() {
    const labels = [];
    const os = env('RUNNER_OS');
    const arch = env('RUNNER_ARCH');
    const name = env('RUNNER_NAME');
    if (os)
        labels.push(os.toLowerCase());
    if (arch)
        labels.push(arch.toLowerCase());
    if (name.toLowerCase().includes('self-hosted'))
        labels.push('self-hosted');
    if (name.toLowerCase().startsWith('github actions'))
        labels.push('github-hosted');
    return labels.length ? labels : undefined;
}
function githubContextToSignalSubject(ctx, extra) {
    return {
        repository: ctx.repository,
        commitSha: ctx.commitSha,
        workflowRunId: ctx.workflowRunId,
        runAttempt: ctx.runAttempt,
        jobName: ctx.jobName,
        ...extra,
    };
}
/** Safe subset of GitHub env vars for evidence files (no secrets). */
function safeGithubEnvRecord() {
    const keys = [
        'GITHUB_ACTION',
        'GITHUB_ACTIONS',
        'GITHUB_ACTOR',
        'GITHUB_API_URL',
        'GITHUB_BASE_REF',
        'GITHUB_EVENT_NAME',
        'GITHUB_JOB',
        'GITHUB_REF',
        'GITHUB_REF_NAME',
        'GITHUB_REF_TYPE',
        'GITHUB_REPOSITORY',
        'GITHUB_REPOSITORY_ID',
        'GITHUB_REPOSITORY_OWNER',
        'GITHUB_RUN_ATTEMPT',
        'GITHUB_RUN_ID',
        'GITHUB_RUN_NUMBER',
        'GITHUB_SERVER_URL',
        'GITHUB_SHA',
        'GITHUB_WORKFLOW',
        'GITHUB_WORKFLOW_REF',
        'RUNNER_ARCH',
        'GITHUB_ENVIRONMENT',
        'RUNNER_OS',
        'RUNNER_NAME',
    ];
    const out = {};
    for (const k of keys) {
        const v = env(k);
        if (v)
            out[k] = v;
    }
    return out;
}
