// @bridgedai/actions-core/dist/graph-hints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.graphEdge = graphEdge;
exports.repositoryToBranch = repositoryToBranch;
exports.branchToCommit = branchToCommit;
exports.commitToWorkflowRun = commitToWorkflowRun;
exports.workflowRunToRunner = workflowRunToRunner;
exports.pullRequestToCommit = pullRequestToCommit;
exports.commitToArtifact = commitToArtifact;
exports.artifactToContainerImage = artifactToContainerImage;
exports.containerImageToRegistry = containerImageToRegistry;
exports.artifactToSbom = artifactToSbom;
exports.sbomToPackage = sbomToPackage;
exports.packageToDependency = packageToDependency;
exports.artifactToDeployment = artifactToDeployment;
exports.containerImageToDeployment = containerImageToDeployment;
exports.deploymentToEnvironment = deploymentToEnvironment;
exports.deploymentToService = deploymentToService;
exports.deploymentToWorkload = deploymentToWorkload;
exports.workloadToContainerImage = workloadToContainerImage;
exports.artifactToAttestation = artifactToAttestation;
exports.attestationToBuilder = attestationToBuilder;
exports.attestationToSourceCommit = attestationToSourceCommit;
function graphEdge(fromType, fromId, toType, toId, relation) {
    return { fromType, fromId, toType, toId, relation };
}
function repositoryToBranch(repo, branch) {
    return graphEdge('Repository', repo, 'Branch', `${repo}@${branch}`, 'has_branch');
}
function branchToCommit(branch, commitSha) {
    return graphEdge('Branch', branch, 'Commit', commitSha, 'points_to');
}
function commitToWorkflowRun(commitSha, runId) {
    return graphEdge('Commit', commitSha, 'WorkflowRun', runId, 'triggered');
}
function workflowRunToRunner(runId, runnerName) {
    return graphEdge('WorkflowRun', runId, 'Runner', runnerName, 'executed_on');
}
function pullRequestToCommit(prNumber, commitSha) {
    return graphEdge('PullRequest', String(prNumber), 'Commit', commitSha, 'includes');
}
function commitToArtifact(commitSha, digest) {
    return graphEdge('Commit', commitSha, 'Artifact', digest, 'produced');
}
function artifactToContainerImage(artifactDigest, imageDigest) {
    return graphEdge('Artifact', artifactDigest, 'ContainerImage', imageDigest, 'packaged_as');
}
function containerImageToRegistry(imageDigest, registry) {
    return graphEdge('ContainerImage', imageDigest, 'Registry', registry, 'hosted_in');
}
function artifactToSbom(artifactDigest, sbomDigest) {
    return graphEdge('Artifact', artifactDigest, 'SBOM', sbomDigest, 'described_by');
}
function sbomToPackage(sbomDigest, packageId) {
    return graphEdge('SBOM', sbomDigest, 'Package', packageId, 'lists');
}
function packageToDependency(packageId, dependencyId) {
    return graphEdge('Package', packageId, 'Dependency', dependencyId, 'depends_on');
}
function artifactToDeployment(artifactDigest, deploymentId) {
    return graphEdge('Artifact', artifactDigest, 'Deployment', deploymentId, 'deployed_as');
}
function containerImageToDeployment(imageDigest, deploymentId) {
    return graphEdge('ContainerImage', imageDigest, 'Deployment', deploymentId, 'deployed_as');
}
function deploymentToEnvironment(deploymentId, environment) {
    return graphEdge('Deployment', deploymentId, 'Environment', environment, 'targets');
}
function deploymentToService(deploymentId, service) {
    return graphEdge('Deployment', deploymentId, 'Service', service, 'serves');
}
function deploymentToWorkload(deploymentId, workload) {
    return graphEdge('Deployment', deploymentId, 'Workload', workload, 'manages');
}
function workloadToContainerImage(workload, imageDigest) {
    return graphEdge('Workload', workload, 'ContainerImage', imageDigest, 'runs');
}
function artifactToAttestation(artifactDigest, attestationId) {
    return graphEdge('Artifact', artifactDigest, 'Attestation', attestationId, 'attested_by');
}
function attestationToBuilder(attestationId, builderId) {
    return graphEdge('Attestation', attestationId, 'Builder', builderId, 'built_by');
}
function attestationToSourceCommit(attestationId, commitSha) {
    return graphEdge('Attestation', attestationId, 'SourceCommit', commitSha, 'sources_from');
}
