// @bridgedai/actions-core/dist/idempotency.js
"use strict";
/** Deterministic idempotency key helpers for GitHub Actions signals. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.workflowSignalKey = workflowSignalKey;
exports.artifactSignalKey = artifactSignalKey;
exports.ociSignalKey = ociSignalKey;
exports.deploymentSignalKey = deploymentSignalKey;
exports.deploymentRegisteredIdempotencyKey = deploymentRegisteredIdempotencyKey;
exports.runnerUsedSignalKey = runnerUsedSignalKey;
function sanitize(part) {
    return String(part ?? '')
        .trim()
        .replace(/\s+/g, '_')
        .replace(/[^a-zA-Z0-9._/@:-]/g, '-');
}
function workflowSignalKey(ctx, signalType, subjectDigest) {
    const base = [
        'github-actions',
        sanitize(ctx.ownerRepo),
        sanitize(ctx.workflowRunId),
        sanitize(ctx.runAttempt),
        sanitize(ctx.jobName),
        sanitize(signalType),
    ].join(':');
    if (subjectDigest) {
        return `${base}:${sanitize(subjectDigest.replace(/^sha256:/i, ''))}`;
    }
    return base;
}
function artifactSignalKey(ownerRepo, workflowRunId, signalType, artifactDigest) {
    return [
        'github-actions',
        sanitize(ownerRepo),
        sanitize(workflowRunId),
        sanitize(signalType),
        sanitize(artifactDigest.replace(/^sha256:/i, '')),
    ].join(':');
}
function ociSignalKey(registry, image, digest, signalType) {
    return ['oci', sanitize(registry), sanitize(image), sanitize(digest.replace(/^sha256:/i, '')), sanitize(signalType)].join(':');
}
function deploymentSignalKey(environment, service, digest, deploymentId, signalType) {
    return [
        'deployment',
        sanitize(environment),
        sanitize(service),
        sanitize(digest.replace(/^sha256:/i, '')),
        sanitize(deploymentId),
        sanitize(signalType),
    ].join(':');
}
function deploymentRegisteredIdempotencyKey(params) {
    const rawDigest = params.artifactDigest ?? params.imageDigest ?? '';
    const hex = rawDigest.replace(/^sha256:/i, '').trim();
    const digestPart = hex ? `:${hex}:` : params.deploymentName || 'none';
    return [
        'deployment-registered',
        sanitize(params.repository),
        sanitize(params.workflowRunId),
        sanitize(params.runAttempt),
        sanitize(params.environment),
        sanitize(digestPart),
    ].join(':');
}
function runnerUsedSignalKey(ctx) {
    const jobSegment = ctx.workflowJobId ? `job-${sanitize(ctx.workflowJobId)}` : sanitize(ctx.jobName);
    return [
        'github-actions',
        sanitize(ctx.ownerRepo),
        sanitize(ctx.workflowRunId),
        sanitize(ctx.runAttempt),
        jobSegment,
        'runner.used',
    ].join(':');
}
