// @bridgedai/actions-core/dist/public-api.js
"use strict";
/**
 * Public BridgedAI HTTP API surface for https://api.bridgedai.io/
 * @see private/actions-dev/docs/API_COMPATIBILITY.md
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BridgedApiHttpError = exports.BridgedPublicApiPaths = exports.DEFAULT_BRIDGED_API_BASE = exports.isSigstoreCryptoSupported = exports.isDeploymentRegistrationSupported = exports.getActionsCapabilities = exports.BRIDGED_ACTIONS_CAPABILITIES = void 0;
exports.joinBridgedApiUrl = joinBridgedApiUrl;
exports.resolveBridgedApiBase = resolveBridgedApiBase;
exports.normalizeGateDecisionForOutput = normalizeGateDecisionForOutput;
exports.bearerRequestJson = bearerRequestJson;
exports.exchangeGithubOidcToken = exchangeGithubOidcToken;
exports.evaluateReleaseGateWithBearer = evaluateReleaseGateWithBearer;
exports.getCapabilitiesWithBearer = getCapabilitiesWithBearer;
const capabilities_1 = require("./capabilities");
var capabilities_2 = require("./capabilities");
Object.defineProperty(exports, "BRIDGED_ACTIONS_CAPABILITIES", { enumerable: true, get: function () { return capabilities_2.BRIDGED_ACTIONS_CAPABILITIES; } });
Object.defineProperty(exports, "getActionsCapabilities", { enumerable: true, get: function () { return capabilities_2.getActionsCapabilities; } });
Object.defineProperty(exports, "isDeploymentRegistrationSupported", { enumerable: true, get: function () { return capabilities_2.isDeploymentRegistrationSupported; } });
Object.defineProperty(exports, "isSigstoreCryptoSupported", { enumerable: true, get: function () { return capabilities_2.isSigstoreCryptoSupported; } });
exports.DEFAULT_BRIDGED_API_BASE = 'https://api.bridgedai.io';
/** Paths are relative to the API host root (no `/api` prefix). */
exports.BridgedPublicApiPaths = {
    /** @deprecated Backend removed GET /v1/actions/capabilities — use getActionsCapabilities() instead. */
    capabilities: '/v1/actions/capabilities',
    usersMe: '/v1/users/me',
    buildsResolve: '/v1/builds/resolve',
    buildsEnsure: '/v1/builds/ensure',
    ingestSbom: '/v1/ingest/sbom',
    ingestAttestation: '/v1/ingest/attestation',
    ingestEvidence: '/v1/ingest/evidence',
    releaseGateEvaluate: '/v1/enforcement/release-gate/evaluate',
    verifySupplyChain: (artifactId) => `/v1/artifacts/${encodeURIComponent(artifactId)}/verify-supply-chain`,
    buildSupplyChainVerification: (buildId) => `/v1/builds/${encodeURIComponent(buildId)}/supply-chain-verification`,
    /** GitHub Actions OIDC workload token exchange. */
    authWorkloadExchange: '/v1/auth/workload/exchange',
    /** Canonical GitHub Actions OIDC exchange (preferred). */
    authGithubActionsOidcExchange: '/v1/auth/oidc/github-actions/exchange',
    /** @deprecated Alias — use authGithubActionsOidcExchange */
    authGithubOidcExchange: '/v1/auth/oidc/github-actions/exchange',
};
function joinBridgedApiUrl(baseUrl, path) {
    const base = baseUrl.replace(/\/+$/, '');
    const p = path.startsWith('/') ? path : `/${path}`;
    return `${base}${p}`;
}
function resolveBridgedApiBase(raw, fieldName = 'api-base', defaultWhenEmpty = exports.DEFAULT_BRIDGED_API_BASE) {
    const trimmed = String(raw ?? '').trim();
    if (!trimmed)
        return defaultWhenEmpty.replace(/\/+$/, '');
    return trimmed.replace(/\/+$/, '');
}
/** Map public API decision (`allow` | `warn` | `block`) to action output style. */
function normalizeGateDecisionForOutput(decision) {
    const d = String(decision ?? '').trim().toLowerCase();
    if (d === 'allow')
        return 'ALLOW';
    if (d === 'warn')
        return 'WARN';
    if (d === 'block')
        return 'BLOCK';
    const u = String(decision ?? '').trim().toUpperCase();
    if (u === 'ALLOW' || u === 'WARN' || u === 'BLOCK')
        return u;
    throw new Error(`Unexpected release gate decision: ${decision}`);
}
async function sleep(ms) {
    await new Promise((r) => setTimeout(r, ms));
}
async function bearerRequestJson(opts) {
    const url = joinBridgedApiUrl(opts.baseUrl, opts.path);
    const timeoutMs = opts.timeoutMs ?? 30_000;
    const allowRetry = opts.allowRetry ?? opts.method === 'GET';
    const maxAttempts = allowRetry ? 3 : 1;
    let lastErr;
    for (let attempt = 0; attempt < maxAttempts; attempt++) {
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), timeoutMs);
        try {
            const headers = {
                accept: 'application/json',
                authorization: `Bearer ${opts.accessToken}`,
                ...(opts.orgId ? { 'x-org-id': opts.orgId } : {}),
            };
            const init = { method: opts.method, headers, signal: controller.signal };
            if (opts.body !== undefined && opts.method === 'POST') {
                headers['content-type'] = 'application/json';
                init.body = JSON.stringify(opts.body);
            }
            const res = await fetch(url, init);
            const text = await res.text();
            let parsed;
            if (text) {
                try {
                    parsed = JSON.parse(text);
                }
                catch {
                    parsed = { raw: text };
                }
            }
            if (res.ok)
                return parsed;
            const retryable = res.status === 502 || res.status === 503 || res.status === 504;
            if (allowRetry && retryable && attempt < maxAttempts - 1) {
                await sleep(250 * 2 ** attempt);
                continue;
            }
            const detail = parsed && typeof parsed === 'object' && 'detail' in parsed
                ? String(parsed.detail)
                : text.slice(0, 500);
            throw new BridgedApiHttpError(res.status, `BridgedAI API ${res.status}: ${detail || res.statusText}`);
        }
        catch (e) {
            lastErr = e;
            if (e instanceof Error && e.name === 'AbortError') {
                throw new Error(`BridgedAI request timed out after ${timeoutMs}ms (${opts.method} ${opts.path})`);
            }
            throw e;
        }
        finally {
            clearTimeout(timer);
        }
    }
    throw lastErr instanceof Error ? lastErr : new Error(String(lastErr));
}
class BridgedApiHttpError extends Error {
    status;
    constructor(status, message) {
        super(message);
        this.status = status;
        this.name = 'BridgedApiHttpError';
    }
}
exports.BridgedApiHttpError = BridgedApiHttpError;
async function exchangeGithubOidcToken(baseUrl, body, timeoutMs = 30_000) {
    const orgId = String(body.organizationId ?? body.orgId ?? body.tenant ?? '').trim();
    const exchangeBody = {
        token: body.gitHubOidcToken,
        oidcJwt: body.gitHubOidcToken,
        provider: 'github',
        organizationId: orgId,
        repository: body.repository,
        workflowRunId: body.workflowRunId,
        commitSha: body.commitSha,
        audience: body.audience,
    };
    const paths = [
        exports.BridgedPublicApiPaths.authGithubActionsOidcExchange,
        exports.BridgedPublicApiPaths.authWorkloadExchange,
    ];
    let lastStatus = 0;
    let lastDetail = '';
    for (const path of paths) {
        const url = joinBridgedApiUrl(baseUrl, path);
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), timeoutMs);
        try {
            const res = await fetch(url, {
                method: 'POST',
                headers: {
                    accept: 'application/json',
                    'content-type': 'application/json',
                    ...(orgId ? { 'x-org-id': orgId } : {}),
                },
                body: JSON.stringify(exchangeBody),
                signal: controller.signal,
            });
            const text = await res.text();
            if (res.status === 404 && path === paths[0]) {
                lastStatus = 404;
                lastDetail = text.slice(0, 200);
                continue;
            }
            if (!res.ok) {
                throw new BridgedApiHttpError(res.status, `BridgedAI OIDC exchange failed (${res.status}): ${text.slice(0, 500)}`);
            }
            const parsed = JSON.parse(text);
            const accessToken = String(parsed.accessToken ?? parsed.access_token ?? '').trim();
            const expiresIn = Number(parsed.expiresIn ?? 0);
            return {
                ...parsed,
                access_token: accessToken,
                accessToken,
                tenant_id: String(parsed.organizationId ?? parsed.orgId ?? orgId),
                tenantId: String(parsed.organizationId ?? parsed.orgId ?? orgId),
                session_id: String(parsed.identityId ?? parsed.session_id ?? parsed.sessionId ?? ''),
                sessionId: String(parsed.identityId ?? parsed.session_id ?? parsed.sessionId ?? ''),
                expires_at: parsed.expires_at ??
                    parsed.expiresAt ??
                    (expiresIn > 0 ? new Date(Date.now() + expiresIn * 1000).toISOString() : undefined),
            };
        }
        catch (e) {
            if (e instanceof Error && e.name === 'AbortError') {
                throw new Error(`BridgedAI OIDC exchange timed out after ${timeoutMs}ms`);
            }
            if (e instanceof BridgedApiHttpError)
                throw e;
            throw e;
        }
        finally {
            clearTimeout(timer);
        }
    }
    throw new BridgedApiHttpError(lastStatus || 404, `BridgedAI OIDC exchange endpoint unavailable${lastDetail ? `: ${lastDetail}` : ''}`);
}
/** Release gate evaluate using a BridgedAI JWT from OIDC exchange (Bearer auth). */
async function evaluateReleaseGateWithBearer(baseUrl, accessToken, body, timeoutMs = 60_000) {
    const orgId = String(body.orgId ?? '').trim();
    return bearerRequestJson({
        baseUrl,
        accessToken,
        orgId: orgId || undefined,
        method: 'POST',
        path: exports.BridgedPublicApiPaths.releaseGateEvaluate,
        body,
        timeoutMs,
        allowRetry: false,
    });
}
async function getCapabilitiesWithBearer(_baseUrl, _accessToken, _timeoutMs = 30_000) {
    void _baseUrl;
    void _accessToken;
    void _timeoutMs;
    return (0, capabilities_1.getActionsCapabilities)();
}
