// @bridgedai/actions-core/dist/runner-metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.inferRunnerHostedType = inferRunnerHostedType;
exports.buildRunnerUsedPayload = buildRunnerUsedPayload;
exports.buildRunnerUsedNormalizedPayload = buildRunnerUsedNormalizedPayload;
function normLabels(raw) {
    if (!Array.isArray(raw))
        return [];
    return raw.map((l) => String(l ?? '').trim()).filter(Boolean);
}
function inferRunnerHostedType(input) {
    const labels = normLabels(input.labels);
    const lowerLabels = labels.map((l) => l.toLowerCase());
    const name = (input.runnerName ?? '').toLowerCase();
    if (lowerLabels.includes('self-hosted') || name.includes('self-hosted')) {
        return { hostedType: 'self_hosted', selfHosted: true, ephemeral: 'unknown' };
    }
    if (name.startsWith('github actions') ||
        lowerLabels.includes('github-hosted') ||
        (name.includes('hosted') && !lowerLabels.includes('self-hosted'))) {
        return { hostedType: 'github_hosted', selfHosted: false, ephemeral: true };
    }
    return { hostedType: 'unknown', selfHosted: false, ephemeral: 'unknown' };
}
function buildRunnerUsedPayload(input) {
    const labels = normLabels(input.labels);
    const inferred = inferRunnerHostedType({ runnerName: input.runnerName, labels });
    const providerId = `github:${input.runnerName ?? 'unknown'}:${input.runnerGroup ?? ''}:${[...labels].sort().join(',')}`;
    return {
        id: providerId,
        name: input.runnerName,
        labels,
        os: input.runnerOs,
        architecture: input.runnerArch,
        runnerGroup: input.runnerGroup,
        hostedType: inferred.hostedType,
        ephemeral: inferred.ephemeral,
        workflowJobId: input.workflowJobId,
    };
}
function buildRunnerUsedNormalizedPayload(input) {
    return {
        signalType: 'runner.used',
        organizationId: input.organizationId,
        source: 'github_actions',
        repository: input.repository,
        workflowRunId: input.workflowRunId,
        workflowJobId: input.workflowJobId,
        commitSha: input.commitSha,
        branch: input.branch,
        runner: input.runner,
        observedAt: input.observedAt ?? new Date().toISOString(),
    };
}
