// @bridgedai/actions-core/dist/runner-context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildRunnerUsedSignalKey = void 0;
exports.normalizeRunnerContext = normalizeRunnerContext;
exports.sanitizeRunnerMetadata = sanitizeRunnerMetadata;
exports.buildRunnerRegisterPayload = buildRunnerRegisterPayload;
exports.dedupeSignalEnvelope = dedupeSignalEnvelope;
exports.appendSignalEnvelope = appendSignalEnvelope;
const github_context_1 = require("./github-context");
const idempotency_1 = require("./idempotency");
Object.defineProperty(exports, "buildRunnerUsedSignalKey", { enumerable: true, get: function () { return idempotency_1.runnerUsedSignalKey; } });
const redaction_1 = require("./redaction");
const signal_writer_1 = require("./signal-writer");
const signal_envelope_1 = require("./signal-envelope");
/** Resolve GitHub workflow context and idempotency fields for runner.used signals. */
function normalizeRunnerContext(opts) {
    const ghCtx = (0, github_context_1.extractGitHubContext)(opts?.changedFiles);
    const workflowJobId = String(opts?.workflowJobId ?? ghCtx.workflowJobId ?? '').trim() || undefined;
    if (workflowJobId)
        ghCtx.workflowJobId = workflowJobId;
    return {
        ghCtx,
        idempotency: {
            ownerRepo: ghCtx.repository,
            workflowRunId: ghCtx.workflowRunId,
            runAttempt: ghCtx.runAttempt,
            jobName: ghCtx.jobName,
            workflowJobId,
        },
    };
}
const REGISTER_PAYLOAD_KEYS = new Set([
    'repository',
    'workflow',
    'workflowRunId',
    'workflowJobId',
    'jobName',
    'runnerName',
    'runnerOs',
    'runnerArchitecture',
    'runnerGroup',
    'labels',
    'commitSha',
    'branch',
    'actor',
    'runAttempt',
    'buildId',
    'projectId',
    'correlationId',
    'dryRun',
]);
/** Strip unknown fields and redact sensitive values from runner registration payloads. */
function sanitizeRunnerMetadata(payload) {
    const out = {};
    for (const [key, value] of Object.entries(payload)) {
        if (!REGISTER_PAYLOAD_KEYS.has(key))
            continue;
        if (value === undefined || value === null || value === '')
            continue;
        out[key] = (0, redaction_1.redactUnknown)(value);
    }
    return out;
}
/** Build POST /v1/runners/register body from GitHub context (backend RegisterRunnerDto fields only). */
function buildRunnerRegisterPayload(input) {
    const { ghCtx } = input;
    const workflowJobId = String(input.workflowJobIdOverride ?? ghCtx.workflowJobId ?? '').trim() || undefined;
    return sanitizeRunnerMetadata({
        repository: ghCtx.repository,
        workflow: ghCtx.workflowName,
        workflowRunId: ghCtx.workflowRunId,
        workflowJobId,
        jobName: ghCtx.jobName,
        runnerName: ghCtx.runnerName,
        runnerOs: ghCtx.runnerOs,
        runnerArchitecture: ghCtx.runnerArch,
        runnerGroup: ghCtx.runnerGroup,
        labels: ghCtx.runnerLabels,
        commitSha: ghCtx.commitSha,
        branch: ghCtx.branch,
        actor: ghCtx.actor,
        runAttempt: Number(ghCtx.runAttempt) || 1,
        buildId: input.buildId,
        projectId: input.projectId,
        correlationId: input.correlationId,
        dryRun: input.dryRun,
    });
}
/** Returns true when an envelope with the same idempotency key already exists in the file. */
async function dedupeSignalEnvelope(filePath, signal) {
    (0, signal_envelope_1.validateSignalEnvelope)(signal);
    const existing = await (0, signal_writer_1.readSignalsFromFile)(filePath);
    return existing.some((s) => s.idempotencyKey === signal.idempotencyKey);
}
/**
 * Append a signal envelope; skip when the canonical idempotency key already exists.
 * Returns whether a new line was written.
 */
async function appendSignalEnvelope(filePath, signal) {
    (0, signal_envelope_1.validateSignalEnvelope)(signal);
    if (await dedupeSignalEnvelope(filePath, signal)) {
        return 'deduplicated';
    }
    await (0, signal_writer_1.appendSignalToFile)(filePath, signal);
    return 'appended';
}
