// @bridgedai/actions-core/dist/runner-metadata.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const runner_metadata_1 = require("../src/runner-metadata");
(0, vitest_1.describe)('runner-metadata', () => {
    (0, vitest_1.it)('infers github hosted runner', () => {
        const inferred = (0, runner_metadata_1.inferRunnerHostedType)({ runnerName: 'GitHub Actions 2', labels: ['ubuntu-latest'] });
        (0, vitest_1.expect)(inferred.hostedType).toBe('github_hosted');
        (0, vitest_1.expect)(inferred.ephemeral).toBe(true);
        (0, vitest_1.expect)(inferred.selfHosted).toBe(false);
    });
    (0, vitest_1.it)('infers self-hosted runner', () => {
        const inferred = (0, runner_metadata_1.inferRunnerHostedType)({ runnerName: 'my-runner', labels: ['self-hosted', 'linux'] });
        (0, vitest_1.expect)(inferred.hostedType).toBe('self_hosted');
        (0, vitest_1.expect)(inferred.selfHosted).toBe(true);
    });
    (0, vitest_1.it)('handles unknown labels without crashing', () => {
        const payload = (0, runner_metadata_1.buildRunnerUsedPayload)({ labels: undefined, runnerName: 'custom-agent' });
        (0, vitest_1.expect)(payload.labels).toEqual([]);
        (0, vitest_1.expect)(payload.hostedType).toBe('unknown');
    });
    (0, vitest_1.it)('builds runner.used payload without secrets', () => {
        const payload = (0, runner_metadata_1.buildRunnerUsedPayload)({
            runnerName: 'GitHub Actions 2',
            runnerOs: 'Linux',
            runnerArch: 'X64',
            labels: ['ubuntu-latest'],
            workflowJobId: '12345',
        });
        (0, vitest_1.expect)(payload.hostedType).toBe('github_hosted');
        (0, vitest_1.expect)(JSON.stringify(payload)).not.toMatch(/ghs_|ghp_|Bearer /);
    });
    (0, vitest_1.it)('includes workflowJobId in normalized payload when provided', () => {
        const runner = (0, runner_metadata_1.buildRunnerUsedPayload)({ runnerName: 'GitHub Actions 2', labels: ['ubuntu-latest'] });
        const normalized = (0, runner_metadata_1.buildRunnerUsedNormalizedPayload)({
            organizationId: 'org_1',
            repository: 'org/repo',
            workflowRunId: '999',
            workflowJobId: '12345',
            runner,
        });
        (0, vitest_1.expect)(normalized['workflowJobId']).toBe('12345');
        (0, vitest_1.expect)(normalized['signalType']).toBe('runner.used');
        (0, vitest_1.expect)(normalized['source']).toBe('github_actions');
    });
});
