// @bridgedai/actions-core/dist/signal-envelope.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SIGNAL_SCHEMA_VERSION = void 0;
exports.createSignalEnvelope = createSignalEnvelope;
exports.validateSignalEnvelope = validateSignalEnvelope;
const verification_level_1 = require("./verification-level");
exports.SIGNAL_SCHEMA_VERSION = '2026-05-01';
function createSignalEnvelope(opts) {
    const orgId = String(opts.organizationId ?? '').trim();
    if (!orgId)
        throw new Error('organizationId is required for signal envelope');
    const signalType = String(opts.signalType ?? '').trim();
    if (!signalType)
        throw new Error('signalType is required for signal envelope');
    const idempotencyKey = String(opts.idempotencyKey ?? '').trim();
    if (!idempotencyKey)
        throw new Error('idempotencyKey is required for signal envelope');
    const verificationLevel = opts.verificationLevel ?? 'provider_authoritative';
    (0, verification_level_1.assertVerificationLevel)(verificationLevel);
    return {
        organizationId: orgId,
        provider: 'github',
        sourceSystem: 'github-actions',
        signalType,
        schemaVersion: exports.SIGNAL_SCHEMA_VERSION,
        externalId: opts.externalId ?? idempotencyKey,
        idempotencyKey,
        observedAt: opts.observedAt ?? new Date().toISOString(),
        confidence: opts.confidence ?? 'high',
        verificationLevel,
        subject: opts.subject ?? {},
        rawPayload: opts.rawPayload ?? {},
        normalizedPayload: opts.normalizedPayload ?? {},
        evidenceHints: opts.evidenceHints ?? [],
        graphHints: opts.graphHints ?? [],
        metadata: opts.metadata ?? {},
    };
}
function validateSignalEnvelope(envelope) {
    if (!envelope || typeof envelope !== 'object') {
        throw new Error('Signal envelope must be an object');
    }
    const e = envelope;
    const required = [
        'organizationId',
        'provider',
        'sourceSystem',
        'signalType',
        'schemaVersion',
        'externalId',
        'idempotencyKey',
        'observedAt',
        'confidence',
        'verificationLevel',
    ];
    for (const key of required) {
        if (typeof e[key] !== 'string' || !String(e[key]).trim()) {
            throw new Error(`Signal envelope missing or invalid field: ${key}`);
        }
    }
    if (e.provider !== 'github')
        throw new Error('Signal envelope provider must be github');
    if (e.sourceSystem !== 'github-actions') {
        throw new Error('Signal envelope sourceSystem must be github-actions');
    }
    (0, verification_level_1.assertVerificationLevel)(String(e.verificationLevel));
    if (!e.subject || typeof e.subject !== 'object') {
        throw new Error('Signal envelope subject must be an object');
    }
    if (!Array.isArray(e.evidenceHints))
        throw new Error('Signal envelope evidenceHints must be an array');
    if (!Array.isArray(e.graphHints))
        throw new Error('Signal envelope graphHints must be an array');
    if (!e.rawPayload || typeof e.rawPayload !== 'object') {
        throw new Error('Signal envelope rawPayload must be an object');
    }
    if (!e.normalizedPayload || typeof e.normalizedPayload !== 'object') {
        throw new Error('Signal envelope normalizedPayload must be an object');
    }
    return e;
}
