// @bridgedai/actions-core/dist/signal-ingest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ingestSignals = ingestSignals;
const bridgedai_client_1 = require("./bridgedai-client");
const signal_writer_1 = require("./signal-writer");
const BATCH_FORMAT = 'bridgedai-signal-batch-v1';
function buildBatchPayload(deduped, batchId) {
    return {
        type: 'bridgedai_signal_batch',
        schemaVersion: '2026-05-01',
        batchId,
        signalCount: deduped.length,
        signals: deduped,
    };
}
async function tryNativeSignalIngest(client, req, deduped, batchId) {
    try {
        const res = await client.requestJson({
            method: 'POST',
            path: '/v1/signals/ingest',
            body: {
                batchId,
                organizationId: req.orgId,
                source: req.source ?? 'github-actions',
                format: BATCH_FORMAT,
                signals: deduped,
            },
        });
        const failed = Number(res.failed ?? 0);
        return {
            ok: failed === 0,
            parsed: true,
            status: String(res.status ?? 'accepted'),
            evidenceEmitted: Number(res.evidenceCreated ?? deduped.length),
        };
    }
    catch (e) {
        const status = e && typeof e === 'object' && 'status' in e ? Number(e.status) : 0;
        if (status === 404 || status === 501)
            return null;
        throw e;
    }
}
/**

 * Upload canonical signal envelopes.

 * 1. Try POST /v1/signals/ingest (native pipeline).

 * 2. Fall back to POST /v1/ingest/evidence (kind=artifact_metadata, format=bridgedai-signal-batch-v1).

 */
async function ingestSignals(client, req) {
    const deduped = (0, signal_writer_1.dedupeSignalsByIdempotencyKey)(req.signals);
    const batchId = req.batchId ?? `sig-batch-${Date.now()}`;
    const results = [];
    let uploaded = 0;
    let failed = 0;
    let ingestPath = 'artifact_metadata';
    const batchPayload = buildBatchPayload(deduped, batchId);
    const native = await tryNativeSignalIngest(client, req, deduped, batchId);
    if (native) {
        results.push(native);
        ingestPath = 'native';
        if (native.ok)
            uploaded = deduped.length;
        else
            failed = deduped.length;
        return { uploaded, deduped: req.signals.length - deduped.length, failed, batchId, ingestPath, results };
    }
    try {
        const res = await (0, bridgedai_client_1.ingestEvidence)(client, {
            orgId: req.orgId,
            buildId: req.buildId,
            projectId: req.projectId,
            kind: 'artifact_metadata',
            source: req.source ?? 'github-actions',
            format: BATCH_FORMAT,
            payload: batchPayload,
            metadata: {
                idempotencyKey: `signals:${req.orgId}:${req.buildId ?? req.repoFullName ?? 'unknown'}:${batchId}`,
                signalCount: deduped.length,
                repoFullName: req.repoFullName,
            },
        });
        results.push(res);
        if (res.ok)
            uploaded = deduped.length;
        else
            failed = deduped.length;
    }
    catch {
        ingestPath = 'per_signal_fallback';
        for (const signal of deduped) {
            try {
                const res = await (0, bridgedai_client_1.ingestEvidence)(client, {
                    orgId: req.orgId,
                    buildId: req.buildId,
                    projectId: req.projectId,
                    kind: 'artifact_metadata',
                    source: req.source ?? 'github-actions',
                    format: 'bridgedai-signal-v1',
                    payload: { type: 'bridgedai_signal', signal },
                    metadata: {
                        idempotencyKey: signal.idempotencyKey,
                        signalType: signal.signalType,
                    },
                });
                results.push(res);
                if (res.ok)
                    uploaded += 1;
                else
                    failed += 1;
            }
            catch {
                failed += 1;
            }
        }
    }
    return {
        uploaded,
        deduped: req.signals.length - deduped.length,
        failed,
        batchId,
        ingestPath,
        results,
    };
}
