// @bridgedai/actions-core/dist/signal-writer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_SIGNALS_BATCH_FILE = exports.DEFAULT_SIGNALS_FILE = void 0;
exports.ensureDirForFile = ensureDirForFile;
exports.readSignalsFromFile = readSignalsFromFile;
exports.appendSignalToFile = appendSignalToFile;
exports.writeSignalsBatchFile = writeSignalsBatchFile;
exports.mergeSignalsToBatch = mergeSignalsToBatch;
exports.dedupeSignalsByIdempotencyKey = dedupeSignalsByIdempotencyKey;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const signal_envelope_1 = require("./signal-envelope");
exports.DEFAULT_SIGNALS_FILE = '.bridgedai/signals.jsonl';
exports.DEFAULT_SIGNALS_BATCH_FILE = '.bridgedai/signals-batch.json';
async function ensureDirForFile(filePath) {
    const dir = path.dirname(path.resolve(filePath));
    await fs.promises.mkdir(dir, { recursive: true });
}
async function readSignalsFromFile(filePath) {
    const abs = path.resolve(filePath);
    if (!fs.existsSync(abs))
        return [];
    const raw = await fs.promises.readFile(abs, 'utf8');
    const trimmed = raw.trim();
    if (!trimmed)
        return [];
    if (abs.endsWith('.jsonl')) {
        const lines = trimmed.split('\n').filter((l) => l.trim());
        return lines.map((line, i) => {
            try {
                return (0, signal_envelope_1.validateSignalEnvelope)(JSON.parse(line));
            }
            catch (e) {
                throw new Error(`Invalid signal at ${abs}:${i + 1}: ${e instanceof Error ? e.message : String(e)}`);
            }
        });
    }
    const parsed = JSON.parse(trimmed);
    if (Array.isArray(parsed)) {
        return parsed.map((item, i) => {
            try {
                return (0, signal_envelope_1.validateSignalEnvelope)(item);
            }
            catch (e) {
                throw new Error(`Invalid signal at ${abs}[${i}]: ${e instanceof Error ? e.message : String(e)}`);
            }
        });
    }
    if (parsed && typeof parsed === 'object' && Array.isArray(parsed.signals)) {
        return parsed.signals.map((item, i) => {
            try {
                return (0, signal_envelope_1.validateSignalEnvelope)(item);
            }
            catch (e) {
                throw new Error(`Invalid signal in batch at ${abs}[${i}]: ${e instanceof Error ? e.message : String(e)}`);
            }
        });
    }
    return [(0, signal_envelope_1.validateSignalEnvelope)(parsed)];
}
async function appendSignalToFile(filePath, signal) {
    (0, signal_envelope_1.validateSignalEnvelope)(signal);
    await ensureDirForFile(filePath);
    const abs = path.resolve(filePath);
    const line = `${JSON.stringify(signal)}\n`;
    await fs.promises.appendFile(abs, line, { encoding: 'utf8', mode: 0o644 });
}
async function writeSignalsBatchFile(filePath, signals, metadata) {
    for (const s of signals)
        (0, signal_envelope_1.validateSignalEnvelope)(s);
    await ensureDirForFile(filePath);
    const abs = path.resolve(filePath);
    const doc = {
        schemaVersion: '2026-05-01',
        type: 'bridgedai_signal_batch',
        signalCount: signals.length,
        signals,
        metadata: metadata ?? {},
    };
    await fs.promises.writeFile(abs, `${JSON.stringify(doc, null, 2)}\n`, { encoding: 'utf8', mode: 0o644 });
}
async function mergeSignalsToBatch(outputFile, newSignals) {
    const existing = fs.existsSync(path.resolve(outputFile)) ? await readSignalsFromFile(outputFile) : [];
    const byKey = new Map();
    for (const s of existing)
        byKey.set(s.idempotencyKey, s);
    for (const s of newSignals)
        byKey.set(s.idempotencyKey, s);
    const merged = [...byKey.values()];
    await writeSignalsBatchFile(outputFile, merged);
    return merged;
}
function dedupeSignalsByIdempotencyKey(signals) {
    const byKey = new Map();
    for (const s of signals) {
        (0, signal_envelope_1.validateSignalEnvelope)(s);
        byKey.set(s.idempotencyKey, s);
    }
    return [...byKey.values()];
}
